{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "from NIPDToolkit.xlinked_analysis import XLinkedAnalyser\n",
    "from NIPDToolkit.autosomal_analysis import AutosomalAnalyser\n",
    "import yaml\n",
    "import pandas as pd\n",
    "import csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "config = '../development_local.yaml'\n",
    "min_genotype_depth = 20\n",
    "min_distance_between_snps = 200\n",
    "min_snps_per_block = 25\n",
    "min_fetal_fraction =0.0\n",
    "results_dir = '../output/sprt/'\n",
    "worksheet_name = '190201_NB551415_0009_AHF7C7BGX7'\n",
    "x_genes = ['EDA', 'F8', 'F9']\n",
    "autosomal_genes = ['CFTR', 'TSC1', 'TSC2']\n",
    "family_csv_dir = '../output/family_csvs/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_config(config_location):\n",
    "    \"\"\"\n",
    "    Parse the YAML config file.\n",
    "    \"\"\"\n",
    "\n",
    "    with open(config_location, 'r') as stream:\n",
    "\n",
    "        try:\n",
    "            return yaml.safe_load(stream)\n",
    "        except yaml.YAMLError as exc:\n",
    "            print(exc)\n",
    "            raise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "config_dict = parse_config(config)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "families = config_dict['families'].keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Can perform X Linked SPRT for family FAM001\n",
      "Can do Autosomal Analysis\n",
      "Can perform X Linked SPRT for family FAM002\n",
      "Can do Autosomal Analysis\n",
      "Can perform X Linked SPRT for family FAM003\n",
      "Can do Autosomal Analysis\n",
      "Can perform X Linked SPRT for family FAM004\n",
      "Can do Autosomal Analysis\n",
      "Can perform X Linked SPRT for family FAM005\n",
      "Can do Autosomal Analysis\n",
      "Can perform X Linked SPRT for family FAM006\n",
      "Can do Autosomal Analysis\n"
     ]
    }
   ],
   "source": [
    "for family in families:\n",
    "    \n",
    "    members = config_dict['families'][family]['members']\n",
    "    \n",
    "    mother = members['mother']\n",
    "    father = members['father']\n",
    "    plasma = members['plasma']\n",
    "    proband = members['proband']\n",
    "    \n",
    "    if mother != None and plasma != None and proband != None:\n",
    "        \n",
    "        print (f'Can perform X Linked SPRT for family {family}')\n",
    "        \n",
    "        for gene in x_genes:\n",
    "            \n",
    "            family_csv = f'{family_csv_dir}/190201_NB551415_0009_AHF7C7BGX7_all_chr_qfiltered_anno_selected_{family}.csv'\n",
    "            \n",
    "            my_x_linked_analyser = XLinkedAnalyser(family_csv=family_csv,\n",
    "                                      maternal_sample= mother,\n",
    "                                      proband_sample= proband,\n",
    "                                      plasma_sample=plasma,\n",
    "                                      gene=gene,\n",
    "                                      min_genotype_depth=min_genotype_depth,\n",
    "                                      min_distance_between_snps=min_distance_between_snps,\n",
    "                                      min_snps_per_block=min_snps_per_block,\n",
    "                                      min_fetal_fraction=min_fetal_fraction)\n",
    "            \n",
    "            my_x_linked_analyser.run_analysis()\n",
    "            \n",
    "            my_x_linked_analyser.xlinked_df_fwd.to_csv(f'{results_dir}/{worksheet_name}_{family}_{gene}_raw_x_linked_fwd.csv', index=False)\n",
    "            my_x_linked_analyser.xlinked_df_rev.to_csv(f'{results_dir}/{worksheet_name}_{family}_{gene}_raw_x_linked_rev.csv', index=False)\n",
    "            \n",
    "            blocks_fwd = my_x_linked_analyser.xlinked_df_fwd_haplotype_blocks\n",
    "            blocks_fwd_df = pd.DataFrame(blocks_fwd, columns=['block_id', 'start', 'end', 'n_snps', 'haplotype', 'chomosome', 'gene'])\n",
    "            blocks_fwd_df.to_csv(f'{results_dir}/{worksheet_name}_{family}_{gene}_haplotype_blocks_x_linked_fwd.csv', index=False)\n",
    "            \n",
    "            blocks_rev = my_x_linked_analyser.xlinked_df_rev_haplotype_blocks\n",
    "            blocks_rev_df = pd.DataFrame(blocks_rev, columns=['block_id', 'start', 'end', 'n_snps', 'haplotype', 'chomosome', 'gene'])\n",
    "            blocks_rev_df.to_csv(f'{results_dir}/{worksheet_name}_{family}_{gene}_haplotype_blocks_x_linked_rev.csv', index=False)\n",
    "            \n",
    "\n",
    "            with open(f'{results_dir}/{worksheet_name}_{family}_{gene}_x_linked_info.csv', 'w') as csvfile:\n",
    "                spamwriter = csv.writer(csvfile, delimiter=',')\n",
    "                spamwriter.writerow(['mother:',mother])\n",
    "                spamwriter.writerow(['father:',father])\n",
    "                spamwriter.writerow(['plasma:',plasma])\n",
    "                spamwriter.writerow(['proband:',proband])\n",
    "                spamwriter.writerow(['gene:',gene])\n",
    "                spamwriter.writerow(['min_genotype_depth:',my_x_linked_analyser.min_genotype_depth])\n",
    "                spamwriter.writerow(['min_distance_between_snps:',my_x_linked_analyser.min_distance_between_snps])\n",
    "                spamwriter.writerow(['min_snps_per_block:',my_x_linked_analyser.min_snps_per_block])\n",
    "                spamwriter.writerow(['min_fetal_fraction:', my_x_linked_analyser.min_fetal_fraction])\n",
    "                spamwriter.writerow(['fetal_fraction:',my_x_linked_analyser.fetal_fraction])\n",
    "                spamwriter.writerow(['initial_variant_count:',my_x_linked_analyser.initial_variant_count])\n",
    "                spamwriter.writerow(['after_depth_filter_variant_count:',my_x_linked_analyser.after_depth_filter_variant_count])\n",
    "                spamwriter.writerow(['after_gene_filter_variant_count:',my_x_linked_analyser.after_gene_filter_variant_count])\n",
    "                spamwriter.writerow(['after_distance_filter_variant_count:',my_x_linked_analyser.after_distance_filter_variant_count])\n",
    "                spamwriter.writerow(['informative_snp_count:',my_x_linked_analyser.informative_snp_count])\n",
    "                spamwriter.writerow(['mean_snp_depth:',my_x_linked_analyser.mean_snp_depth])\n",
    "                \n",
    "                \n",
    "    if mother != None and plasma != None and proband != None and father != None:     \n",
    "        \n",
    "        print ('Can do Autosomal Analysis')\n",
    "         \n",
    "        for gene in autosomal_genes:\n",
    "            \n",
    "            family_csv = f'{family_csv_dir}/190201_NB551415_0009_AHF7C7BGX7_all_chr_qfiltered_anno_selected_{family}.csv'\n",
    "            \n",
    "            my_autosomal_analyser = AutosomalAnalyser(\n",
    "                family_csv = family_csv,\n",
    "                maternal_sample = mother,\n",
    "                paternal_sample = father,\n",
    "                proband_sample = proband,\n",
    "                plasma_sample = plasma,\n",
    "                gene = gene,\n",
    "                min_genotype_depth=min_genotype_depth,\n",
    "                min_distance_between_snps=min_distance_between_snps,\n",
    "                min_snps_per_block=min_snps_per_block,\n",
    "                min_fetal_fraction=min_fetal_fraction)\n",
    "            \n",
    "            \n",
    "            my_autosomal_analyser.run_analysis()\n",
    "            \n",
    "            \n",
    "            my_autosomal_analyser.type4a_df_fwd.to_csv(f'{results_dir}/{worksheet_name}_{family}_{gene}_raw_type4a_fwd.csv', index=False)\n",
    "            my_autosomal_analyser.type4a_df_rev.to_csv(f'{results_dir}/{worksheet_name}_{family}_{gene}_raw_type4a_rev.csv', index=False)\n",
    "            \n",
    "            my_autosomal_analyser.type4b_df_fwd.to_csv(f'{results_dir}/{worksheet_name}_{family}_{gene}_raw_type4b_fwd.csv', index=False)\n",
    "            my_autosomal_analyser.type4b_df_rev.to_csv(f'{results_dir}/{worksheet_name}_{family}_{gene}_raw_type4b_rev.csv', index=False)\n",
    "            \n",
    "            my_autosomal_analyser.type3_df_fwd.to_csv(f'{results_dir}/{worksheet_name}_{family}_{gene}_raw_type3_fwd.csv', index=False)\n",
    "            my_autosomal_analyser.type3_df_rev.to_csv(f'{results_dir}/{worksheet_name}_{family}_{gene}_raw_type3_rev.csv', index=False)\n",
    "            \n",
    "            blocks_fwd = my_autosomal_analyser.type4a_df_fwd_haplotype_blocks\n",
    "            blocks_fwd_df = pd.DataFrame(blocks_fwd, columns=['block_id', 'start', 'end', 'n_snps', 'haplotype', 'chomosome', 'gene'])\n",
    "            blocks_fwd_df.to_csv(f'{results_dir}/{worksheet_name}_{family}_{gene}_haplotype_blocks_type4a_fwd.csv', index=False)\n",
    "            \n",
    "            blocks_rev = my_autosomal_analyser.type4a_df_rev_haplotype_blocks\n",
    "            blocks_rev_df = pd.DataFrame(blocks_rev, columns=['block_id', 'start', 'end', 'n_snps', 'haplotype', 'chomosome', 'gene'])\n",
    "            blocks_rev_df.to_csv(f'{results_dir}/{worksheet_name}_{family}_{gene}_haplotype_blocks_type4a_rev.csv', index=False)\n",
    "            \n",
    "            blocks_fwd = my_autosomal_analyser.type4b_df_fwd_haplotype_blocks\n",
    "            blocks_fwd_df = pd.DataFrame(blocks_fwd, columns=['block_id', 'start', 'end', 'n_snps', 'haplotype', 'chomosome', 'gene'])\n",
    "            blocks_fwd_df.to_csv(f'{results_dir}/{worksheet_name}_{family}_{gene}_haplotype_blocks_type4b_fwd.csv', index=False)\n",
    "            \n",
    "            blocks_rev = my_autosomal_analyser.type4b_df_rev_haplotype_blocks\n",
    "            blocks_rev_df = pd.DataFrame(blocks_rev, columns=['block_id', 'start', 'end', 'n_snps', 'haplotype', 'chomosome', 'gene'])\n",
    "            blocks_rev_df.to_csv(f'{results_dir}/{worksheet_name}_{family}_{gene}_haplotype_blocks_type4b_rev.csv', index=False)\n",
    "            \n",
    "            \n",
    "            blocks_fwd = my_autosomal_analyser.type3_df_fwd_haplotype_blocks\n",
    "            blocks_fwd_df = pd.DataFrame(blocks_fwd, columns=['block_id', 'start', 'end', 'n_snps', 'haplotype', 'chomosome', 'gene'])\n",
    "            blocks_fwd_df.to_csv(f'{results_dir}/{worksheet_name}_{family}_{gene}_haplotype_blocks_type3_fwd.csv', index=False)\n",
    "            \n",
    "            blocks_rev = my_autosomal_analyser.type3_df_rev_haplotype_blocks\n",
    "            blocks_rev_df = pd.DataFrame(blocks_rev, columns=['block_id', 'start', 'end', 'n_snps', 'haplotype', 'chomosome', 'gene'])\n",
    "            blocks_rev_df.to_csv(f'{results_dir}/{worksheet_name}_{family}_{gene}_haplotype_blocks_type3_rev.csv', index=False)\n",
    "            \n",
    "            \n",
    "            \n",
    "            \n",
    "            \n",
    "            with open(f'{results_dir}/{worksheet_name}_{family}_{gene}_autosomal_info.csv', 'w') as csvfile:\n",
    "                spamwriter = csv.writer(csvfile, delimiter=',')\n",
    "                spamwriter.writerow(['mother:',mother])\n",
    "                spamwriter.writerow(['father:',father])\n",
    "                spamwriter.writerow(['plasma:',plasma])\n",
    "                spamwriter.writerow(['proband:',proband])\n",
    "                spamwriter.writerow(['gene:',gene])\n",
    "                spamwriter.writerow(['min_genotype_depth:',my_autosomal_analyser.min_genotype_depth])\n",
    "                spamwriter.writerow(['min_distance_between_snps:',my_autosomal_analyser.min_distance_between_snps])\n",
    "                spamwriter.writerow(['min_snps_per_block:',my_autosomal_analyser.min_snps_per_block])\n",
    "                spamwriter.writerow(['min_fetal_fraction:', my_autosomal_analyser.min_fetal_fraction])\n",
    "                spamwriter.writerow(['min_ks_p_value:', my_autosomal_analyser.min_ks_p_value])\n",
    "                spamwriter.writerow(['fetal_fraction:',my_autosomal_analyser.fetal_fraction])\n",
    "                spamwriter.writerow(['initial_variant_count:',my_autosomal_analyser.initial_variant_count])\n",
    "                spamwriter.writerow(['after_depth_filter_variant_count:',my_autosomal_analyser.after_depth_filter_variant_count])\n",
    "                spamwriter.writerow(['after_distance_filter_variant_count:',my_autosomal_analyser.after_distance_filter_variant_count])\n",
    "                spamwriter.writerow(['after_gene_filter_variant_count:',my_autosomal_analyser.after_gene_filter_variant_count])\n",
    "                spamwriter.writerow(['mean_snp_depth:',my_autosomal_analyser.mean_snp_depth])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
