{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 406,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "from collections import OrderedDict\n",
    "import matplotlib.patches as patches\n",
    "sns.set_style(\"white\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 407,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Performing SPRT X-Linked Analysis\n",
      "The maternal sample is:\t14M01888\n",
      "The proband sample is:\t14M02709\n",
      "The maternal sample is:\tF2-10\n",
      "Selected gene for analysis is:\tF8\n",
      "Minimum SNP genotype depth is set to:\t30\n",
      "Minimum distance between SNPs is set to:\t200\n",
      "Minimum SNPs per haplotype block is set to:\t25\n"
     ]
    }
   ],
   "source": [
    "# Set some variables\n",
    "\n",
    "family_csv = '../../181129_NB551415_0005_AHWW37BGX7/family_csvs/181129_NB551415_0005_AHWW37BGX7_all_chr_qfiltered_anno_selected_FAM002.csv'\n",
    "\n",
    "maternal_sample = '14M01888'\n",
    "proband_sample = '14M02709'\n",
    "plasma_sample = 'F2-10'\n",
    "\n",
    "gene = 'F8'\n",
    "\n",
    "min_genotype_depth = 30\n",
    "min_distance_between_snps = 200\n",
    "min_snps_per_block = 25\n",
    "\n",
    "print ('Performing SPRT X-Linked Analysis')\n",
    "print (f'The maternal sample is:\\t{maternal_sample}')\n",
    "print (f'The proband sample is:\\t{proband_sample}')\n",
    "print (f'The maternal sample is:\\t{plasma_sample}')\n",
    "print (f'Selected gene for analysis is:\\t{gene}')\n",
    "print (f'Minimum SNP genotype depth is set to:\\t{min_genotype_depth}')\n",
    "print (f'Minimum distance between SNPs is set to:\\t{min_distance_between_snps}')\n",
    "print (f'Minimum SNPs per haplotype block is set to:\\t{min_snps_per_block}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 408,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(family_csv, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 409,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>ID</th>\n",
       "      <th>GENE</th>\n",
       "      <th>14M01888.GT</th>\n",
       "      <th>14M01888.AD</th>\n",
       "      <th>14M01888.DP</th>\n",
       "      <th>14M02709.GT</th>\n",
       "      <th>14M02709.AD</th>\n",
       "      <th>14M02709.DP</th>\n",
       "      <th>F2-10.GT</th>\n",
       "      <th>F2-10.AD</th>\n",
       "      <th>F2-10.DP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7</td>\n",
       "      <td>114498885</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>T/T</td>\n",
       "      <td>130,0</td>\n",
       "      <td>130</td>\n",
       "      <td>T/T</td>\n",
       "      <td>120,0</td>\n",
       "      <td>120</td>\n",
       "      <td>T/T</td>\n",
       "      <td>293,0</td>\n",
       "      <td>293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7</td>\n",
       "      <td>114499932</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>C/C</td>\n",
       "      <td>1,331</td>\n",
       "      <td>332</td>\n",
       "      <td>C/C</td>\n",
       "      <td>0,419</td>\n",
       "      <td>419</td>\n",
       "      <td>C/C</td>\n",
       "      <td>0,878</td>\n",
       "      <td>878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7</td>\n",
       "      <td>114537363</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>A/A</td>\n",
       "      <td>2,400</td>\n",
       "      <td>402</td>\n",
       "      <td>A/A</td>\n",
       "      <td>0,477</td>\n",
       "      <td>477</td>\n",
       "      <td>A/A</td>\n",
       "      <td>0,885</td>\n",
       "      <td>885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7</td>\n",
       "      <td>114539933</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>C/C</td>\n",
       "      <td>1,195</td>\n",
       "      <td>196</td>\n",
       "      <td>C/C</td>\n",
       "      <td>0,155</td>\n",
       "      <td>159</td>\n",
       "      <td>C/C</td>\n",
       "      <td>0,432</td>\n",
       "      <td>443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7</td>\n",
       "      <td>114559777</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>G/G</td>\n",
       "      <td>146,0</td>\n",
       "      <td>146</td>\n",
       "      <td>G/G</td>\n",
       "      <td>152,0</td>\n",
       "      <td>152</td>\n",
       "      <td>G/G</td>\n",
       "      <td>378,0</td>\n",
       "      <td>378</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM        POS REF ALT ID  GENE 14M01888.GT 14M01888.AD  14M01888.DP  \\\n",
       "0     7  114498885   T   C  .  CFTR         T/T       130,0          130   \n",
       "1     7  114499932   T   C  .  CFTR         C/C       1,331          332   \n",
       "2     7  114537363   T   A  .  CFTR         A/A       2,400          402   \n",
       "3     7  114539933   A   C  .  CFTR         C/C       1,195          196   \n",
       "4     7  114559777   G   A  .  CFTR         G/G       146,0          146   \n",
       "\n",
       "  14M02709.GT 14M02709.AD  14M02709.DP F2-10.GT F2-10.AD  F2-10.DP  \n",
       "0         T/T       120,0          120      T/T    293,0       293  \n",
       "1         C/C       0,419          419      C/C    0,878       878  \n",
       "2         A/A       0,477          477      A/A    0,885       885  \n",
       "3         C/C       0,155          159      C/C    0,432       443  \n",
       "4         G/G       152,0          152      G/G    378,0       378  "
      ]
     },
     "execution_count": 409,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 410,
   "metadata": {},
   "outputs": [],
   "source": [
    "def order_columns(df):\n",
    "    \"\"\"\n",
    "    Ensure that the columns are ordered as:\n",
    "    \n",
    "    CHROM, POS, REF, ALT, ID, GENE, maternal, proband, plasma\n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    df = df[['CHROM',\n",
    "             'POS',\n",
    "             'REF',\n",
    "             'ALT',\n",
    "             'ID',\n",
    "             'GENE',\n",
    "             maternal_sample +'.GT',\n",
    "             maternal_sample +'.AD',\n",
    "             maternal_sample +'.DP',\n",
    "             proband_sample +'.GT',\n",
    "             proband_sample +'.AD',\n",
    "             proband_sample +'.DP',            \n",
    "             plasma_sample +'.GT',\n",
    "             plasma_sample +'.AD',\n",
    "             plasma_sample +'.DP',\n",
    "            ]]\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 411,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = order_columns(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 412,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>ID</th>\n",
       "      <th>GENE</th>\n",
       "      <th>14M01888.GT</th>\n",
       "      <th>14M01888.AD</th>\n",
       "      <th>14M01888.DP</th>\n",
       "      <th>14M02709.GT</th>\n",
       "      <th>14M02709.AD</th>\n",
       "      <th>14M02709.DP</th>\n",
       "      <th>F2-10.GT</th>\n",
       "      <th>F2-10.AD</th>\n",
       "      <th>F2-10.DP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7</td>\n",
       "      <td>114498885</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>T/T</td>\n",
       "      <td>130,0</td>\n",
       "      <td>130</td>\n",
       "      <td>T/T</td>\n",
       "      <td>120,0</td>\n",
       "      <td>120</td>\n",
       "      <td>T/T</td>\n",
       "      <td>293,0</td>\n",
       "      <td>293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7</td>\n",
       "      <td>114499932</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>C/C</td>\n",
       "      <td>1,331</td>\n",
       "      <td>332</td>\n",
       "      <td>C/C</td>\n",
       "      <td>0,419</td>\n",
       "      <td>419</td>\n",
       "      <td>C/C</td>\n",
       "      <td>0,878</td>\n",
       "      <td>878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7</td>\n",
       "      <td>114537363</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>A/A</td>\n",
       "      <td>2,400</td>\n",
       "      <td>402</td>\n",
       "      <td>A/A</td>\n",
       "      <td>0,477</td>\n",
       "      <td>477</td>\n",
       "      <td>A/A</td>\n",
       "      <td>0,885</td>\n",
       "      <td>885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7</td>\n",
       "      <td>114539933</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>C/C</td>\n",
       "      <td>1,195</td>\n",
       "      <td>196</td>\n",
       "      <td>C/C</td>\n",
       "      <td>0,155</td>\n",
       "      <td>159</td>\n",
       "      <td>C/C</td>\n",
       "      <td>0,432</td>\n",
       "      <td>443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7</td>\n",
       "      <td>114559777</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>G/G</td>\n",
       "      <td>146,0</td>\n",
       "      <td>146</td>\n",
       "      <td>G/G</td>\n",
       "      <td>152,0</td>\n",
       "      <td>152</td>\n",
       "      <td>G/G</td>\n",
       "      <td>378,0</td>\n",
       "      <td>378</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM        POS REF ALT ID  GENE 14M01888.GT 14M01888.AD  14M01888.DP  \\\n",
       "0     7  114498885   T   C  .  CFTR         T/T       130,0          130   \n",
       "1     7  114499932   T   C  .  CFTR         C/C       1,331          332   \n",
       "2     7  114537363   T   A  .  CFTR         A/A       2,400          402   \n",
       "3     7  114539933   A   C  .  CFTR         C/C       1,195          196   \n",
       "4     7  114559777   G   A  .  CFTR         G/G       146,0          146   \n",
       "\n",
       "  14M02709.GT 14M02709.AD  14M02709.DP F2-10.GT F2-10.AD  F2-10.DP  \n",
       "0         T/T       120,0          120      T/T    293,0       293  \n",
       "1         C/C       0,419          419      C/C    0,878       878  \n",
       "2         A/A       0,477          477      A/A    0,885       885  \n",
       "3         C/C       0,155          159      C/C    0,432       443  \n",
       "4         G/G       152,0          152      G/G    378,0       378  "
      ]
     },
     "execution_count": 412,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 413,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fix_column_names(columns):\n",
    "    \"\"\"\n",
    "    Change column names to valid python variable names by replacing '.' with '_'\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    fixed_columns = []\n",
    "    \n",
    "    for column_name in columns:\n",
    "        \n",
    "        fixed_column_name = column_name.replace('.', '_')\n",
    "        \n",
    "        fixed_columns.append(fixed_column_name)\n",
    "    \n",
    "    return fixed_columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 344,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.columns = fix_column_names(df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 345,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The initial variant count in all genes is:\t14473\n"
     ]
    }
   ],
   "source": [
    "variant_count = df.shape[0]\n",
    "\n",
    "print (f'The initial variant count in all genes is:\\t{variant_count}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 346,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The variant count after depth filtering @ GT depth < 30 in all genes is:\t14336\n"
     ]
    }
   ],
   "source": [
    "# Filter out low depth variants - drop entire row if any genotype depths < cut off\n",
    "\n",
    "df = df[(df[maternal_sample + '_DP'] > min_genotype_depth) &\n",
    "       (df[proband_sample + '_DP'] > min_genotype_depth) &\n",
    "       (df[plasma_sample + '_DP'] > min_genotype_depth)]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "variant_count = df.shape[0]\n",
    "print (f'The variant count after depth filtering @ GT depth < {min_genotype_depth} in all genes is:\\t{variant_count}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 347,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The variant count in gene F8 after depth and gene filtering is :\t1920\n"
     ]
    }
   ],
   "source": [
    "# Filter out variants not in selected gene.\n",
    "\n",
    "df = df[df['GENE'] == gene]\n",
    "\n",
    "variant_count = df.shape[0]\n",
    "print (f'The variant count in gene {gene} after depth and gene filtering is :\\t{variant_count}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 348,
   "metadata": {},
   "outputs": [],
   "source": [
    "def genotype_sample(df, sample_name):\n",
    "    \"\"\"\n",
    "    Return whether a sample is HOM or HET\n",
    "    \n",
    "    Add a NA if sample is ./. or 1/. or something else wierd\n",
    "    \"\"\"\n",
    "    \n",
    "    gt = df[sample_name+'_GT']\n",
    "    \n",
    "    if '/' in gt:\n",
    "        \n",
    "        gt = gt.split('/')\n",
    "    \n",
    "    elif '|' in gt:\n",
    "        \n",
    "        gt = gt.split('|')\n",
    "        \n",
    "    else:\n",
    "        \n",
    "        raise ValueError(f\"Found a wierd genotype in variant {df['CHROM']} {df['POS']}\")\n",
    "        \n",
    "    if (gt[0] not in 'ATGC') or (gt[1] not in 'ATGC'):\n",
    "        \n",
    "        return 'NA'\n",
    "        \n",
    "    if gt[0] == gt[1]:\n",
    "        \n",
    "        return 'HOM'\n",
    "    else:\n",
    "        return 'HET'\n",
    "    \n",
    "    \n",
    "def get_allele_depth_ref(df, sample_name):\n",
    "    \"\"\"\n",
    "    For a given sample split the AD column into AD for ref and alts\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    ad = df[sample_name+'_AD']\n",
    "    \n",
    "    ad = ad.split(',')\n",
    "    \n",
    "    return ad[0]\n",
    "    \n",
    "def get_allele_depth_alt(df, sample_name):\n",
    "    \"\"\"\n",
    "    For a given sample split the AD column into AD for ref and alts\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    ad = df[sample_name+'_AD']\n",
    "    \n",
    "    ad = ad.split(',')\n",
    "    \n",
    "    return ad[1]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 349,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Apply genotyping function\n",
    "\n",
    "df[maternal_sample + '_call'] = df.apply(genotype_sample, axis=1, sample_name = maternal_sample )\n",
    "df[proband_sample + '_call'] = df.apply(genotype_sample, axis=1, sample_name = proband_sample )\n",
    "df[plasma_sample + '_call'] = df.apply(genotype_sample, axis=1, sample_name = plasma_sample )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 350,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Apply genotype ref depth function\n",
    "\n",
    "df[maternal_sample + '_AD_ref'] = df.apply(get_allele_depth_ref, axis=1, sample_name = maternal_sample )\n",
    "df[proband_sample + '_AD_ref'] = df.apply(get_allele_depth_ref, axis=1, sample_name = proband_sample )\n",
    "df[plasma_sample + '_AD_ref'] = df.apply(get_allele_depth_ref, axis=1, sample_name = plasma_sample )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 351,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Apply genotype alt depth function\n",
    "\n",
    "df[maternal_sample + '_AD_alt'] = df.apply(get_allele_depth_alt, axis=1, sample_name = maternal_sample )\n",
    "df[proband_sample + '_AD_alt'] = df.apply(get_allele_depth_alt, axis=1, sample_name = proband_sample )\n",
    "df[plasma_sample + '_AD_alt'] = df.apply(get_allele_depth_alt, axis=1, sample_name = plasma_sample )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 352,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The informative variant count for gene (F8) after all filtering is :\t668\n"
     ]
    }
   ],
   "source": [
    "df = df[df[maternal_sample + '_call'] =='HET']\n",
    "df = df[df[proband_sample + '_call'] =='HOM']\n",
    "\n",
    "variant_count = df.shape[0]\n",
    "print(f'The informative variant count for gene ({gene}) after all filtering is :\\t{variant_count}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 353,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_hapA_counts(df, proband_sample, plasma_sample):\n",
    "    \"\"\"\n",
    "    count the number of reads in the plasma sample which match the affected haplotype\n",
    "    \n",
    "    e.g. how many of the reads in the plasma match the genotype of the affected son?\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    # Get the genotype of the proband \n",
    "    gt = df[proband_sample+'_GT']\n",
    "    \n",
    "    if '/' in gt:\n",
    "        \n",
    "        gt = gt.split('/')\n",
    "    \n",
    "    elif '|' in gt:\n",
    "        \n",
    "        gt = gt.split('|')\n",
    "        \n",
    "    else:\n",
    "        \n",
    "        raise ValueError(f\"Found a wierd genotype in variant {df['CHROM']} {df['POS']}\")\n",
    "    \n",
    "    if gt[0] != gt[1]:\n",
    "        \n",
    "        raise ValueError(f\"Found a wierd genotype in variant {df['CHROM']} {df['POS']}\")\n",
    "\n",
    "    gt = gt[0]\n",
    "    \n",
    "    # is the affected haplotype ref or alt?\n",
    "    \n",
    "    affected_haplotype = ''\n",
    "    \n",
    "    if gt == df['REF']:\n",
    "        \n",
    "        affected_haplotype = 'REF'\n",
    "        \n",
    "    elif gt == df['ALT']:\n",
    "        \n",
    "        affected_haplotype = 'ALT'\n",
    "        \n",
    "    # if the affected haplotype is the REF then get the ref AD for the plasma sample\n",
    "        \n",
    "    if affected_haplotype == 'REF':\n",
    "        \n",
    "        return df[plasma_sample+'_AD_ref']\n",
    "    \n",
    "    elif affected_haplotype == 'ALT':\n",
    "        \n",
    "        return df[plasma_sample+'_AD_alt']\n",
    "    \n",
    "def get_hapB_counts(df, proband_sample, plasma_sample):\n",
    "    \"\"\"\n",
    "    count the number of reads in the plasma sample which match the affected haplotype\n",
    "    \n",
    "    e.g. how many of the reads in the plasma match the genotype of the affected son?\n",
    "    \n",
    "    \"\"\"\n",
    "                         \n",
    "    gt = df[proband_sample+'_GT']\n",
    "    \n",
    "    if '/' in gt:\n",
    "        \n",
    "        gt = gt.split('/')\n",
    "    \n",
    "    elif '|' in gt:\n",
    "        \n",
    "        gt = gt.split('|')\n",
    "        \n",
    "    else:\n",
    "        \n",
    "        raise ValueError(f\"Found a wierd genotype in variant {df['CHROM']} {df['POS']}\")\n",
    "    \n",
    "    if gt[0] != gt[1]:\n",
    "        \n",
    "        raise ValueError(f\"Found a wierd genotype in variant {df['CHROM']} {df['POS']}\")\n",
    "\n",
    "    gt = gt[0]\n",
    "    \n",
    "    # is the affected haplotype ref or alt?\n",
    "    \n",
    "    affected_haplotype = ''\n",
    "    \n",
    "    if gt == df['REF']:\n",
    "        \n",
    "        affected_haplotype = 'REF'\n",
    "        \n",
    "    elif gt == df['ALT']:\n",
    "        \n",
    "        affected_haplotype = 'ALT'\n",
    "        \n",
    "    # if the affected haplotype is the REF then get the ref AD for the plasma sample\n",
    "        \n",
    "    if affected_haplotype == 'REF':\n",
    "        \n",
    "        return df[plasma_sample+'_AD_alt']\n",
    "    \n",
    "    elif affected_haplotype == 'ALT':\n",
    "        \n",
    "        return df[plasma_sample+'_AD_ref']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 354,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['hapA_count'] = df.apply(get_hapA_counts, axis=1, proband_sample=proband_sample, plasma_sample=plasma_sample)\n",
    "df['hapB_count'] = df.apply(get_hapB_counts, axis=1, proband_sample=proband_sample, plasma_sample=plasma_sample)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>ID</th>\n",
       "      <th>GENE</th>\n",
       "      <th>14M01888_GT</th>\n",
       "      <th>14M01888_AD</th>\n",
       "      <th>14M01888_DP</th>\n",
       "      <th>14M02709_GT</th>\n",
       "      <th>...</th>\n",
       "      <th>14M02709_call</th>\n",
       "      <th>F1-15_call</th>\n",
       "      <th>14M01888_AD_ref</th>\n",
       "      <th>14M02709_AD_ref</th>\n",
       "      <th>F1-15_AD_ref</th>\n",
       "      <th>14M01888_AD_alt</th>\n",
       "      <th>14M02709_AD_alt</th>\n",
       "      <th>F1-15_AD_alt</th>\n",
       "      <th>hapA_count</th>\n",
       "      <th>hapB_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12567</th>\n",
       "      <td>X</td>\n",
       "      <td>148338598</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>G/A</td>\n",
       "      <td>200,188</td>\n",
       "      <td>388</td>\n",
       "      <td>A/A</td>\n",
       "      <td>...</td>\n",
       "      <td>HOM</td>\n",
       "      <td>HET</td>\n",
       "      <td>200</td>\n",
       "      <td>0</td>\n",
       "      <td>323</td>\n",
       "      <td>188</td>\n",
       "      <td>214</td>\n",
       "      <td>374</td>\n",
       "      <td>374</td>\n",
       "      <td>323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12573</th>\n",
       "      <td>X</td>\n",
       "      <td>148343688</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>G/A</td>\n",
       "      <td>164,160</td>\n",
       "      <td>324</td>\n",
       "      <td>A/A</td>\n",
       "      <td>...</td>\n",
       "      <td>HOM</td>\n",
       "      <td>HET</td>\n",
       "      <td>164</td>\n",
       "      <td>0</td>\n",
       "      <td>210</td>\n",
       "      <td>160</td>\n",
       "      <td>184</td>\n",
       "      <td>294</td>\n",
       "      <td>294</td>\n",
       "      <td>210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12582</th>\n",
       "      <td>X</td>\n",
       "      <td>148354706</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>G/A</td>\n",
       "      <td>160,158</td>\n",
       "      <td>318</td>\n",
       "      <td>A/A</td>\n",
       "      <td>...</td>\n",
       "      <td>HOM</td>\n",
       "      <td>HET</td>\n",
       "      <td>160</td>\n",
       "      <td>1</td>\n",
       "      <td>303</td>\n",
       "      <td>158</td>\n",
       "      <td>176</td>\n",
       "      <td>370</td>\n",
       "      <td>370</td>\n",
       "      <td>303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12583</th>\n",
       "      <td>X</td>\n",
       "      <td>148355171</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>G/A</td>\n",
       "      <td>269,242</td>\n",
       "      <td>511</td>\n",
       "      <td>A/A</td>\n",
       "      <td>...</td>\n",
       "      <td>HOM</td>\n",
       "      <td>HET</td>\n",
       "      <td>269</td>\n",
       "      <td>0</td>\n",
       "      <td>364</td>\n",
       "      <td>242</td>\n",
       "      <td>281</td>\n",
       "      <td>444</td>\n",
       "      <td>444</td>\n",
       "      <td>364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12584</th>\n",
       "      <td>X</td>\n",
       "      <td>148355251</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>C/T</td>\n",
       "      <td>297,307</td>\n",
       "      <td>604</td>\n",
       "      <td>T/T</td>\n",
       "      <td>...</td>\n",
       "      <td>HOM</td>\n",
       "      <td>HET</td>\n",
       "      <td>297</td>\n",
       "      <td>0</td>\n",
       "      <td>444</td>\n",
       "      <td>307</td>\n",
       "      <td>335</td>\n",
       "      <td>526</td>\n",
       "      <td>526</td>\n",
       "      <td>444</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHROM        POS REF ALT ID GENE 14M01888_GT 14M01888_AD  14M01888_DP  \\\n",
       "12567     X  148338598   G   A  .   F8         G/A     200,188          388   \n",
       "12573     X  148343688   G   A  .   F8         G/A     164,160          324   \n",
       "12582     X  148354706   G   A  .   F8         G/A     160,158          318   \n",
       "12583     X  148355171   G   A  .   F8         G/A     269,242          511   \n",
       "12584     X  148355251   C   T  .   F8         C/T     297,307          604   \n",
       "\n",
       "      14M02709_GT    ...     14M02709_call  F1-15_call 14M01888_AD_ref  \\\n",
       "12567         A/A    ...               HOM         HET             200   \n",
       "12573         A/A    ...               HOM         HET             164   \n",
       "12582         A/A    ...               HOM         HET             160   \n",
       "12583         A/A    ...               HOM         HET             269   \n",
       "12584         T/T    ...               HOM         HET             297   \n",
       "\n",
       "      14M02709_AD_ref  F1-15_AD_ref 14M01888_AD_alt 14M02709_AD_alt  \\\n",
       "12567               0           323             188             214   \n",
       "12573               0           210             160             184   \n",
       "12582               1           303             158             176   \n",
       "12583               0           364             242             281   \n",
       "12584               0           444             307             335   \n",
       "\n",
       "      F1-15_AD_alt hapA_count hapB_count  \n",
       "12567          374        374        323  \n",
       "12573          294        294        210  \n",
       "12582          370        370        303  \n",
       "12583          444        444        364  \n",
       "12584          526        526        444  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 355,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 356,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['hapA_count'] = pd.to_numeric(df['hapA_count'])\n",
    "df['hapB_count'] = pd.to_numeric(df['hapB_count'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 383,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_fetal_fraction_x_linked(cHapA, cHapB):\n",
    "    \"\"\"\n",
    "    Use the formula from the Birmingham paper to calculate the fetal \\\n",
    "    fraction for X-linked data.\n",
    "    \n",
    "    https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4864947/\n",
    "    \n",
    "    cffDNA % = [(|cHapA – cHapB| * 2) / (cHapA + cHapB + |cHapA – cHapB|)] * 100\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    fetal_fraction = (abs(cHapA-cHapB)*2) / (cHapA + cHapB + abs(cHapA-cHapB)) *100\n",
    "    \n",
    "    return fetal_fraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 384,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19.7686330360043"
      ]
     },
     "execution_count": 384,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_fetal_fraction_x_linked(df['hapA_count'].sum(), df['hapB_count'].sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 359,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(668, 26)"
      ]
     },
     "execution_count": 359,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 360,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_snps_by_distance(df, min_distance):\n",
    "    \"\"\"\n",
    "    filter snps so there is a minimum distance between of a certain number of basepairs.\n",
    "    \n",
    "    assumes df is sorted in order of position.\n",
    "    \n",
    "    The standard as per most SPRT/RHDO papers is a distance of 200.\n",
    "    \n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    snps_to_keep = []\n",
    "    \n",
    "    last_position = 0\n",
    "    \n",
    "    for snp in df.itertuples():\n",
    "        \n",
    "        position =  snp.POS\n",
    "        \n",
    "        if (position - last_position) > min_distance:\n",
    "            \n",
    "            snps_to_keep.append(snp)\n",
    "            \n",
    "            last_position = position\n",
    "            \n",
    "    return snps_to_keep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 361,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 = pd.DataFrame(filter_snps_by_distance(df, 200))\n",
    "df2.drop('Index', inplace=True,axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 362,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>ID</th>\n",
       "      <th>GENE</th>\n",
       "      <th>_7</th>\n",
       "      <th>_8</th>\n",
       "      <th>_9</th>\n",
       "      <th>_10</th>\n",
       "      <th>...</th>\n",
       "      <th>_17</th>\n",
       "      <th>_18</th>\n",
       "      <th>_19</th>\n",
       "      <th>_20</th>\n",
       "      <th>_21</th>\n",
       "      <th>_22</th>\n",
       "      <th>_23</th>\n",
       "      <th>_24</th>\n",
       "      <th>hapA_count</th>\n",
       "      <th>hapB_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>X</td>\n",
       "      <td>148338598</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>G/A</td>\n",
       "      <td>200,188</td>\n",
       "      <td>388</td>\n",
       "      <td>A/A</td>\n",
       "      <td>...</td>\n",
       "      <td>HOM</td>\n",
       "      <td>HET</td>\n",
       "      <td>200</td>\n",
       "      <td>0</td>\n",
       "      <td>323</td>\n",
       "      <td>188</td>\n",
       "      <td>214</td>\n",
       "      <td>374</td>\n",
       "      <td>374</td>\n",
       "      <td>323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>X</td>\n",
       "      <td>148343688</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>G/A</td>\n",
       "      <td>164,160</td>\n",
       "      <td>324</td>\n",
       "      <td>A/A</td>\n",
       "      <td>...</td>\n",
       "      <td>HOM</td>\n",
       "      <td>HET</td>\n",
       "      <td>164</td>\n",
       "      <td>0</td>\n",
       "      <td>210</td>\n",
       "      <td>160</td>\n",
       "      <td>184</td>\n",
       "      <td>294</td>\n",
       "      <td>294</td>\n",
       "      <td>210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>X</td>\n",
       "      <td>148354706</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>G/A</td>\n",
       "      <td>160,158</td>\n",
       "      <td>318</td>\n",
       "      <td>A/A</td>\n",
       "      <td>...</td>\n",
       "      <td>HOM</td>\n",
       "      <td>HET</td>\n",
       "      <td>160</td>\n",
       "      <td>1</td>\n",
       "      <td>303</td>\n",
       "      <td>158</td>\n",
       "      <td>176</td>\n",
       "      <td>370</td>\n",
       "      <td>370</td>\n",
       "      <td>303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>X</td>\n",
       "      <td>148355171</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>G/A</td>\n",
       "      <td>269,242</td>\n",
       "      <td>511</td>\n",
       "      <td>A/A</td>\n",
       "      <td>...</td>\n",
       "      <td>HOM</td>\n",
       "      <td>HET</td>\n",
       "      <td>269</td>\n",
       "      <td>0</td>\n",
       "      <td>364</td>\n",
       "      <td>242</td>\n",
       "      <td>281</td>\n",
       "      <td>444</td>\n",
       "      <td>444</td>\n",
       "      <td>364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>X</td>\n",
       "      <td>148355553</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>C/A</td>\n",
       "      <td>224,160</td>\n",
       "      <td>384</td>\n",
       "      <td>A/A</td>\n",
       "      <td>...</td>\n",
       "      <td>HOM</td>\n",
       "      <td>HET</td>\n",
       "      <td>224</td>\n",
       "      <td>0</td>\n",
       "      <td>333</td>\n",
       "      <td>160</td>\n",
       "      <td>180</td>\n",
       "      <td>372</td>\n",
       "      <td>372</td>\n",
       "      <td>333</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM        POS REF ALT ID GENE   _7       _8   _9  _10    ...      _17  \\\n",
       "0     X  148338598   G   A  .   F8  G/A  200,188  388  A/A    ...      HOM   \n",
       "1     X  148343688   G   A  .   F8  G/A  164,160  324  A/A    ...      HOM   \n",
       "2     X  148354706   G   A  .   F8  G/A  160,158  318  A/A    ...      HOM   \n",
       "3     X  148355171   G   A  .   F8  G/A  269,242  511  A/A    ...      HOM   \n",
       "4     X  148355553   C   A  .   F8  C/A  224,160  384  A/A    ...      HOM   \n",
       "\n",
       "   _18  _19 _20  _21  _22  _23  _24 hapA_count hapB_count  \n",
       "0  HET  200   0  323  188  214  374        374        323  \n",
       "1  HET  164   0  210  160  184  294        294        210  \n",
       "2  HET  160   1  303  158  176  370        370        303  \n",
       "3  HET  269   0  364  242  281  444        444        364  \n",
       "4  HET  224   0  333  160  180  372        372        333  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 362,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 363,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Do SPRT both 5' to 3' and reversed\n",
    "# code to actually do SPRT\n",
    "# code to get blocks\n",
    "# code to make plots\n",
    "# since this is a panel and we sequence all the genes including autosomal - can we calculate other stats on that?\n",
    "# for example -error rate and ff using X-linked - although we may not have other samples i.e. dad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 364,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sprt_calculate_d(fetal_fraction):\n",
    "    \"\"\"\n",
    "    Calculate the d value needed for SPRT analysis:\n",
    "    \n",
    "    See the following papers for more information:\n",
    "    \n",
    "    1) https://www.ncbi.nlm.nih.gov/pubmed/21148127\n",
    "    \n",
    "    2) https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1934923/bin/pnas_0705765104_index.html\n",
    "    \n",
    "    The d value is used by the function sprt_perform_test()\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    # q0 is the allele imbalance if the null hypothesis is correct i.e. the fetus has the normal allele.\n",
    "    # the 0.5 value is because we would expect to see this much deviation from the 0.5 allele balance seen \\\n",
    "    # in HET SNPs and then some more deviancy depending on the fetal fraction.\n",
    "    \n",
    "    \n",
    "    q0 = 0.5 - (fetal_fraction / 2)\n",
    "    \n",
    "    print (q0)\n",
    "    \n",
    "    # q1 is the allele imbalance if the null hypothesis is incorrect i.e. the fetus has the mutant allele.\n",
    "\n",
    "    q1 = 0.5 + (fetal_fraction / 2)\n",
    "    \n",
    "    print (q1)\n",
    "    \n",
    "    d_value = (1 - q1) / (1 - q0)\n",
    "    \n",
    "    return d_value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 365,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sprt_calculate_g(fetal_fraction):\n",
    "    \"\"\"\n",
    "    Calculate the g value needed for SPRT analysis:\n",
    "    \n",
    "    See the following papers for more information:\n",
    "    \n",
    "    1) https://www.ncbi.nlm.nih.gov/pubmed/21148127\n",
    "    \n",
    "    2) https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1934923/bin/pnas_0705765104_index.html\n",
    "    \n",
    "    The g value is used by the function sprt_perform_test()\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    # q0 is the allele imbalance if the null hypothesis is correct i.e. the fetus has the normal allele.\n",
    "    # the 0.5 value is because we would expect to see this much deviation from the 0.5 allele balance seen \\\n",
    "    # in HET SNPs and then some more deviancy depending on the fetal fraction.\n",
    "    \n",
    "    q0 = 0.5 - (fetal_fraction/2)\n",
    "\n",
    "    # q1 is the allele imbalance if the null hypothesis is incorrect i.e. the fetus has the mutant allele.\n",
    "\n",
    "    q1 = 0.5 + (fetal_fraction/2)\n",
    "\n",
    "    g_value = ( q1*(1 - q0) ) / ( q0*(1 - q1) )\n",
    "    \n",
    "    return g_value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 366,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sprt_calculate_upper_boundry(total_read_count, d_value, g_value):\n",
    "    \"\"\"\n",
    "    Calculate the upper value for the SPRT classification.\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    return (( np.log(1200) / total_read_count ) - np.log(d_value)) / np.log(g_value)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 367,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sprt_calculate_lower_boundry(total_read_count, d_value, g_value):\n",
    "    \"\"\"\n",
    "    Calculate the lower value for the SPRT classification.\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    return (( np.log(float(1) / float(1200) )/ total_read_count ) - np.log(d_value)) / np.log(g_value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 368,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sprt_perform_test(df, d_value, g_value):\n",
    "    \"\"\"\n",
    "    Function to call the status of each snp as hapA, hapB or Unclassified.\n",
    "    \n",
    "    Input:\n",
    "    \n",
    "    df = The processed dataframe containing the X-linked SNP data.\n",
    "    d_value = The value calculated by sprt_calculate_d()\n",
    "    g_value = The value calculated by sprt_calculate_g()\n",
    "    \n",
    "    Output:\n",
    "    \n",
    "    df = A new dataframe containing seven additional rows:\n",
    "    \n",
    "    cumulative_sum_hapa\n",
    "    cumulative_sum_hapb\n",
    "    cumulative_ratio\n",
    "    cumulative_total\n",
    "    upper_boundry\n",
    "    lower_boundry\n",
    "    status\n",
    "    \n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    cumulative_sum_hapa = 0\n",
    "    cumulative_sum_hapb = 0\n",
    "    snp_count = 0\n",
    "    \n",
    "    # Loop through each snp\n",
    "    for snp in df.itertuples():\n",
    "        \n",
    "        # Calculate the cumlative sum\n",
    "        cumulative_sum_hapa = cumulative_sum_hapa + snp.hapA_count\n",
    "        cumulative_sum_hapb = cumulative_sum_hapb + snp.hapB_count\n",
    "        \n",
    "        # Calculate the cumlative allele balance between hapA and hapB\n",
    "        cumulative_ratio = cumulative_sum_hapa / (cumulative_sum_hapa + cumulative_sum_hapb)\n",
    "        \n",
    "        cumulative_total = cumulative_sum_hapa + cumulative_sum_hapb\n",
    "        \n",
    "        # Calculate upper and lower SPRT boundaries\n",
    "        upper_boundry = sprt_calculate_upper_boundry(cumulative_total, d_value, g_value )\n",
    "        \n",
    "        lower_boundry = sprt_calculate_lower_boundry(cumulative_total, d_value, g_value)\n",
    "        \n",
    "        # call blocks either hapA, hapB or Unclassified\n",
    "        \n",
    "        if (cumulative_ratio > upper_boundry) and snp_count >= min_snps_per_block:\n",
    "            \n",
    "            status = 'hapA'\n",
    "            cumulative_sum_hapa = 0\n",
    "            cumulative_sum_hapb = 0\n",
    "            snp_count = 0\n",
    "            \n",
    "            \n",
    "        elif cumulative_ratio < lower_boundry and snp_count >= min_snps_per_block:\n",
    "            \n",
    "            status = 'hapB'\n",
    "            cumulative_sum_hapa = 0\n",
    "            cumulative_sum_hapb = 0\n",
    "            snp_count = 0\n",
    "            \n",
    "        else:\n",
    "            \n",
    "            status = 'Unclassified'\n",
    "            \n",
    "        snp_count = snp_count + 1\n",
    "        \n",
    "        #update dataframe with calculated values\n",
    "        df.at[snp.Index, 'cumulative_sum_hapa'] = cumulative_sum_hapa\n",
    "        df.at[snp.Index, 'cumulative_sum_hapb'] = cumulative_sum_hapb\n",
    "        df.at[snp.Index, 'cumulative_ratio'] = cumulative_ratio\n",
    "        df.at[snp.Index, 'cumulative_total'] = cumulative_total\n",
    "        df.at[snp.Index, 'upper_boundry'] = upper_boundry\n",
    "        df.at[snp.Index, 'lower_boundry'] = lower_boundry\n",
    "        df.at[snp.Index, 'status'] = status\n",
    "        \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 369,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.405\n",
      "0.595\n"
     ]
    }
   ],
   "source": [
    "d_value = sprt_calculate_d(0.19)\n",
    "g_value = sprt_calculate_g(0.19)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 370,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = sprt_perform_test(df2, d_value, g_value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 371,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>ID</th>\n",
       "      <th>GENE</th>\n",
       "      <th>_7</th>\n",
       "      <th>_8</th>\n",
       "      <th>_9</th>\n",
       "      <th>_10</th>\n",
       "      <th>...</th>\n",
       "      <th>_24</th>\n",
       "      <th>hapA_count</th>\n",
       "      <th>hapB_count</th>\n",
       "      <th>cumulative_sum_hapa</th>\n",
       "      <th>cumulative_sum_hapb</th>\n",
       "      <th>cumulative_ratio</th>\n",
       "      <th>cumulative_total</th>\n",
       "      <th>upper_boundry</th>\n",
       "      <th>lower_boundry</th>\n",
       "      <th>status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>X</td>\n",
       "      <td>148338598</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>G/A</td>\n",
       "      <td>200,188</td>\n",
       "      <td>388</td>\n",
       "      <td>A/A</td>\n",
       "      <td>...</td>\n",
       "      <td>374</td>\n",
       "      <td>374</td>\n",
       "      <td>323</td>\n",
       "      <td>374.0</td>\n",
       "      <td>323.0</td>\n",
       "      <td>0.536585</td>\n",
       "      <td>697.0</td>\n",
       "      <td>0.513222</td>\n",
       "      <td>0.486778</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>X</td>\n",
       "      <td>148343688</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>G/A</td>\n",
       "      <td>164,160</td>\n",
       "      <td>324</td>\n",
       "      <td>A/A</td>\n",
       "      <td>...</td>\n",
       "      <td>294</td>\n",
       "      <td>294</td>\n",
       "      <td>210</td>\n",
       "      <td>668.0</td>\n",
       "      <td>533.0</td>\n",
       "      <td>0.556203</td>\n",
       "      <td>1201.0</td>\n",
       "      <td>0.507673</td>\n",
       "      <td>0.492327</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>X</td>\n",
       "      <td>148354706</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>G/A</td>\n",
       "      <td>160,158</td>\n",
       "      <td>318</td>\n",
       "      <td>A/A</td>\n",
       "      <td>...</td>\n",
       "      <td>370</td>\n",
       "      <td>370</td>\n",
       "      <td>303</td>\n",
       "      <td>1038.0</td>\n",
       "      <td>836.0</td>\n",
       "      <td>0.553895</td>\n",
       "      <td>1874.0</td>\n",
       "      <td>0.504918</td>\n",
       "      <td>0.495082</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>X</td>\n",
       "      <td>148355171</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>G/A</td>\n",
       "      <td>269,242</td>\n",
       "      <td>511</td>\n",
       "      <td>A/A</td>\n",
       "      <td>...</td>\n",
       "      <td>444</td>\n",
       "      <td>444</td>\n",
       "      <td>364</td>\n",
       "      <td>1482.0</td>\n",
       "      <td>1200.0</td>\n",
       "      <td>0.552573</td>\n",
       "      <td>2682.0</td>\n",
       "      <td>0.503436</td>\n",
       "      <td>0.496564</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>X</td>\n",
       "      <td>148355553</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>C/A</td>\n",
       "      <td>224,160</td>\n",
       "      <td>384</td>\n",
       "      <td>A/A</td>\n",
       "      <td>...</td>\n",
       "      <td>372</td>\n",
       "      <td>372</td>\n",
       "      <td>333</td>\n",
       "      <td>1854.0</td>\n",
       "      <td>1533.0</td>\n",
       "      <td>0.547387</td>\n",
       "      <td>3387.0</td>\n",
       "      <td>0.502721</td>\n",
       "      <td>0.497279</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>X</td>\n",
       "      <td>148358734</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>T/C</td>\n",
       "      <td>172,176</td>\n",
       "      <td>348</td>\n",
       "      <td>T/T</td>\n",
       "      <td>...</td>\n",
       "      <td>268</td>\n",
       "      <td>335</td>\n",
       "      <td>268</td>\n",
       "      <td>2189.0</td>\n",
       "      <td>1801.0</td>\n",
       "      <td>0.548622</td>\n",
       "      <td>3990.0</td>\n",
       "      <td>0.502310</td>\n",
       "      <td>0.497690</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>X</td>\n",
       "      <td>148361284</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>A/G</td>\n",
       "      <td>369,314</td>\n",
       "      <td>683</td>\n",
       "      <td>G/G</td>\n",
       "      <td>...</td>\n",
       "      <td>498</td>\n",
       "      <td>498</td>\n",
       "      <td>396</td>\n",
       "      <td>2687.0</td>\n",
       "      <td>2197.0</td>\n",
       "      <td>0.550164</td>\n",
       "      <td>4884.0</td>\n",
       "      <td>0.501887</td>\n",
       "      <td>0.498113</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>X</td>\n",
       "      <td>148362146</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>C/T</td>\n",
       "      <td>364,264</td>\n",
       "      <td>628</td>\n",
       "      <td>C/C</td>\n",
       "      <td>...</td>\n",
       "      <td>303</td>\n",
       "      <td>440</td>\n",
       "      <td>303</td>\n",
       "      <td>3127.0</td>\n",
       "      <td>2500.0</td>\n",
       "      <td>0.555714</td>\n",
       "      <td>5627.0</td>\n",
       "      <td>0.501638</td>\n",
       "      <td>0.498362</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>X</td>\n",
       "      <td>148362574</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>T/C</td>\n",
       "      <td>236,201</td>\n",
       "      <td>437</td>\n",
       "      <td>C/C</td>\n",
       "      <td>...</td>\n",
       "      <td>317</td>\n",
       "      <td>317</td>\n",
       "      <td>277</td>\n",
       "      <td>3444.0</td>\n",
       "      <td>2777.0</td>\n",
       "      <td>0.553609</td>\n",
       "      <td>6221.0</td>\n",
       "      <td>0.501481</td>\n",
       "      <td>0.498519</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>X</td>\n",
       "      <td>148369123</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>A/G</td>\n",
       "      <td>258,264</td>\n",
       "      <td>522</td>\n",
       "      <td>A/A</td>\n",
       "      <td>...</td>\n",
       "      <td>356</td>\n",
       "      <td>478</td>\n",
       "      <td>356</td>\n",
       "      <td>3922.0</td>\n",
       "      <td>3133.0</td>\n",
       "      <td>0.555918</td>\n",
       "      <td>7055.0</td>\n",
       "      <td>0.501306</td>\n",
       "      <td>0.498694</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>X</td>\n",
       "      <td>148369607</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>T/C</td>\n",
       "      <td>204,181</td>\n",
       "      <td>385</td>\n",
       "      <td>T/T</td>\n",
       "      <td>...</td>\n",
       "      <td>281</td>\n",
       "      <td>404</td>\n",
       "      <td>281</td>\n",
       "      <td>4326.0</td>\n",
       "      <td>3414.0</td>\n",
       "      <td>0.558915</td>\n",
       "      <td>7740.0</td>\n",
       "      <td>0.501191</td>\n",
       "      <td>0.498809</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>X</td>\n",
       "      <td>148370174</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>G/A</td>\n",
       "      <td>218,211</td>\n",
       "      <td>429</td>\n",
       "      <td>A/A</td>\n",
       "      <td>...</td>\n",
       "      <td>415</td>\n",
       "      <td>415</td>\n",
       "      <td>366</td>\n",
       "      <td>4741.0</td>\n",
       "      <td>3780.0</td>\n",
       "      <td>0.556390</td>\n",
       "      <td>8521.0</td>\n",
       "      <td>0.501082</td>\n",
       "      <td>0.498918</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>X</td>\n",
       "      <td>148532605</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>T/C</td>\n",
       "      <td>155,145</td>\n",
       "      <td>300</td>\n",
       "      <td>T/T</td>\n",
       "      <td>...</td>\n",
       "      <td>236</td>\n",
       "      <td>303</td>\n",
       "      <td>236</td>\n",
       "      <td>5044.0</td>\n",
       "      <td>4016.0</td>\n",
       "      <td>0.556733</td>\n",
       "      <td>9060.0</td>\n",
       "      <td>0.501017</td>\n",
       "      <td>0.498983</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>X</td>\n",
       "      <td>148536132</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>C/T</td>\n",
       "      <td>251,206</td>\n",
       "      <td>457</td>\n",
       "      <td>C/C</td>\n",
       "      <td>...</td>\n",
       "      <td>316</td>\n",
       "      <td>406</td>\n",
       "      <td>316</td>\n",
       "      <td>5450.0</td>\n",
       "      <td>4332.0</td>\n",
       "      <td>0.557146</td>\n",
       "      <td>9782.0</td>\n",
       "      <td>0.500942</td>\n",
       "      <td>0.499058</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>X</td>\n",
       "      <td>148539560</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>A/T</td>\n",
       "      <td>283,252</td>\n",
       "      <td>535</td>\n",
       "      <td>A/A</td>\n",
       "      <td>...</td>\n",
       "      <td>318</td>\n",
       "      <td>395</td>\n",
       "      <td>318</td>\n",
       "      <td>5845.0</td>\n",
       "      <td>4650.0</td>\n",
       "      <td>0.556932</td>\n",
       "      <td>10495.0</td>\n",
       "      <td>0.500878</td>\n",
       "      <td>0.499122</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>X</td>\n",
       "      <td>148544997</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>G/A</td>\n",
       "      <td>113,110</td>\n",
       "      <td>223</td>\n",
       "      <td>G/G</td>\n",
       "      <td>...</td>\n",
       "      <td>197</td>\n",
       "      <td>263</td>\n",
       "      <td>197</td>\n",
       "      <td>6108.0</td>\n",
       "      <td>4847.0</td>\n",
       "      <td>0.557554</td>\n",
       "      <td>10955.0</td>\n",
       "      <td>0.500841</td>\n",
       "      <td>0.499159</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>X</td>\n",
       "      <td>148558352</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>T/C</td>\n",
       "      <td>307,311</td>\n",
       "      <td>630</td>\n",
       "      <td>C/C</td>\n",
       "      <td>...</td>\n",
       "      <td>519</td>\n",
       "      <td>519</td>\n",
       "      <td>427</td>\n",
       "      <td>6627.0</td>\n",
       "      <td>5274.0</td>\n",
       "      <td>0.556844</td>\n",
       "      <td>11901.0</td>\n",
       "      <td>0.500774</td>\n",
       "      <td>0.499226</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>X</td>\n",
       "      <td>148561332</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>C/T</td>\n",
       "      <td>166,173</td>\n",
       "      <td>339</td>\n",
       "      <td>T/T</td>\n",
       "      <td>...</td>\n",
       "      <td>273</td>\n",
       "      <td>273</td>\n",
       "      <td>217</td>\n",
       "      <td>6900.0</td>\n",
       "      <td>5491.0</td>\n",
       "      <td>0.556856</td>\n",
       "      <td>12391.0</td>\n",
       "      <td>0.500744</td>\n",
       "      <td>0.499256</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>X</td>\n",
       "      <td>148563150</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>C/T</td>\n",
       "      <td>80,78</td>\n",
       "      <td>158</td>\n",
       "      <td>T/T</td>\n",
       "      <td>...</td>\n",
       "      <td>152</td>\n",
       "      <td>152</td>\n",
       "      <td>132</td>\n",
       "      <td>7052.0</td>\n",
       "      <td>5623.0</td>\n",
       "      <td>0.556371</td>\n",
       "      <td>12675.0</td>\n",
       "      <td>0.500727</td>\n",
       "      <td>0.499273</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>X</td>\n",
       "      <td>148566218</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>T/C</td>\n",
       "      <td>113,78</td>\n",
       "      <td>191</td>\n",
       "      <td>T/T</td>\n",
       "      <td>...</td>\n",
       "      <td>156</td>\n",
       "      <td>253</td>\n",
       "      <td>156</td>\n",
       "      <td>7305.0</td>\n",
       "      <td>5779.0</td>\n",
       "      <td>0.558315</td>\n",
       "      <td>13084.0</td>\n",
       "      <td>0.500704</td>\n",
       "      <td>0.499296</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>X</td>\n",
       "      <td>148568272</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>A/G</td>\n",
       "      <td>353,288</td>\n",
       "      <td>641</td>\n",
       "      <td>G/G</td>\n",
       "      <td>...</td>\n",
       "      <td>503</td>\n",
       "      <td>503</td>\n",
       "      <td>420</td>\n",
       "      <td>7808.0</td>\n",
       "      <td>6199.0</td>\n",
       "      <td>0.557436</td>\n",
       "      <td>14007.0</td>\n",
       "      <td>0.500658</td>\n",
       "      <td>0.499342</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>X</td>\n",
       "      <td>148592385</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>G/A</td>\n",
       "      <td>300,305</td>\n",
       "      <td>605</td>\n",
       "      <td>G/G</td>\n",
       "      <td>...</td>\n",
       "      <td>400</td>\n",
       "      <td>495</td>\n",
       "      <td>400</td>\n",
       "      <td>8303.0</td>\n",
       "      <td>6599.0</td>\n",
       "      <td>0.557174</td>\n",
       "      <td>14902.0</td>\n",
       "      <td>0.500618</td>\n",
       "      <td>0.499382</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>X</td>\n",
       "      <td>148597990</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>C/G</td>\n",
       "      <td>420,413</td>\n",
       "      <td>833</td>\n",
       "      <td>C/C</td>\n",
       "      <td>...</td>\n",
       "      <td>503</td>\n",
       "      <td>556</td>\n",
       "      <td>503</td>\n",
       "      <td>8859.0</td>\n",
       "      <td>7102.0</td>\n",
       "      <td>0.555040</td>\n",
       "      <td>15961.0</td>\n",
       "      <td>0.500577</td>\n",
       "      <td>0.499423</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>X</td>\n",
       "      <td>148599994</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>T/C</td>\n",
       "      <td>70,81</td>\n",
       "      <td>151</td>\n",
       "      <td>T/T</td>\n",
       "      <td>...</td>\n",
       "      <td>139</td>\n",
       "      <td>201</td>\n",
       "      <td>139</td>\n",
       "      <td>9060.0</td>\n",
       "      <td>7241.0</td>\n",
       "      <td>0.555794</td>\n",
       "      <td>16301.0</td>\n",
       "      <td>0.500565</td>\n",
       "      <td>0.499435</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>X</td>\n",
       "      <td>148605752</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>T/C</td>\n",
       "      <td>336,319</td>\n",
       "      <td>657</td>\n",
       "      <td>T/T</td>\n",
       "      <td>...</td>\n",
       "      <td>387</td>\n",
       "      <td>527</td>\n",
       "      <td>387</td>\n",
       "      <td>9587.0</td>\n",
       "      <td>7628.0</td>\n",
       "      <td>0.556898</td>\n",
       "      <td>17215.0</td>\n",
       "      <td>0.500535</td>\n",
       "      <td>0.499465</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>X</td>\n",
       "      <td>148685623</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>G/C</td>\n",
       "      <td>99,80</td>\n",
       "      <td>179</td>\n",
       "      <td>G/G</td>\n",
       "      <td>...</td>\n",
       "      <td>115</td>\n",
       "      <td>173</td>\n",
       "      <td>115</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.557619</td>\n",
       "      <td>17503.0</td>\n",
       "      <td>0.500527</td>\n",
       "      <td>0.499473</td>\n",
       "      <td>hapA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>X</td>\n",
       "      <td>148846943</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>A/T</td>\n",
       "      <td>45,54</td>\n",
       "      <td>99</td>\n",
       "      <td>A/A</td>\n",
       "      <td>...</td>\n",
       "      <td>117</td>\n",
       "      <td>149</td>\n",
       "      <td>117</td>\n",
       "      <td>149.0</td>\n",
       "      <td>117.0</td>\n",
       "      <td>0.560150</td>\n",
       "      <td>266.0</td>\n",
       "      <td>0.534645</td>\n",
       "      <td>0.465355</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>X</td>\n",
       "      <td>149427266</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>C/T</td>\n",
       "      <td>278,290</td>\n",
       "      <td>568</td>\n",
       "      <td>C/C</td>\n",
       "      <td>...</td>\n",
       "      <td>384</td>\n",
       "      <td>491</td>\n",
       "      <td>384</td>\n",
       "      <td>640.0</td>\n",
       "      <td>501.0</td>\n",
       "      <td>0.560911</td>\n",
       "      <td>1141.0</td>\n",
       "      <td>0.508077</td>\n",
       "      <td>0.491923</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>X</td>\n",
       "      <td>149429373</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>A/G</td>\n",
       "      <td>227,241</td>\n",
       "      <td>468</td>\n",
       "      <td>A/A</td>\n",
       "      <td>...</td>\n",
       "      <td>316</td>\n",
       "      <td>424</td>\n",
       "      <td>316</td>\n",
       "      <td>1064.0</td>\n",
       "      <td>817.0</td>\n",
       "      <td>0.565657</td>\n",
       "      <td>1881.0</td>\n",
       "      <td>0.504899</td>\n",
       "      <td>0.495101</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>X</td>\n",
       "      <td>149434268</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>F8</td>\n",
       "      <td>T/C</td>\n",
       "      <td>249,241</td>\n",
       "      <td>490</td>\n",
       "      <td>T/T</td>\n",
       "      <td>...</td>\n",
       "      <td>339</td>\n",
       "      <td>488</td>\n",
       "      <td>339</td>\n",
       "      <td>1552.0</td>\n",
       "      <td>1156.0</td>\n",
       "      <td>0.573117</td>\n",
       "      <td>2708.0</td>\n",
       "      <td>0.503403</td>\n",
       "      <td>0.496597</td>\n",
       "      <td>Unclassified</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM        POS REF ALT ID GENE   _7       _8   _9  _10      ...       \\\n",
       "0      X  148338598   G   A  .   F8  G/A  200,188  388  A/A      ...        \n",
       "1      X  148343688   G   A  .   F8  G/A  164,160  324  A/A      ...        \n",
       "2      X  148354706   G   A  .   F8  G/A  160,158  318  A/A      ...        \n",
       "3      X  148355171   G   A  .   F8  G/A  269,242  511  A/A      ...        \n",
       "4      X  148355553   C   A  .   F8  C/A  224,160  384  A/A      ...        \n",
       "5      X  148358734   T   C  .   F8  T/C  172,176  348  T/T      ...        \n",
       "6      X  148361284   A   G  .   F8  A/G  369,314  683  G/G      ...        \n",
       "7      X  148362146   C   T  .   F8  C/T  364,264  628  C/C      ...        \n",
       "8      X  148362574   T   C  .   F8  T/C  236,201  437  C/C      ...        \n",
       "9      X  148369123   A   G  .   F8  A/G  258,264  522  A/A      ...        \n",
       "10     X  148369607   T   C  .   F8  T/C  204,181  385  T/T      ...        \n",
       "11     X  148370174   G   A  .   F8  G/A  218,211  429  A/A      ...        \n",
       "12     X  148532605   T   C  .   F8  T/C  155,145  300  T/T      ...        \n",
       "13     X  148536132   C   T  .   F8  C/T  251,206  457  C/C      ...        \n",
       "14     X  148539560   A   T  .   F8  A/T  283,252  535  A/A      ...        \n",
       "15     X  148544997   G   A  .   F8  G/A  113,110  223  G/G      ...        \n",
       "16     X  148558352   T   C  .   F8  T/C  307,311  630  C/C      ...        \n",
       "17     X  148561332   C   T  .   F8  C/T  166,173  339  T/T      ...        \n",
       "18     X  148563150   C   T  .   F8  C/T    80,78  158  T/T      ...        \n",
       "19     X  148566218   T   C  .   F8  T/C   113,78  191  T/T      ...        \n",
       "20     X  148568272   A   G  .   F8  A/G  353,288  641  G/G      ...        \n",
       "21     X  148592385   G   A  .   F8  G/A  300,305  605  G/G      ...        \n",
       "22     X  148597990   C   G  .   F8  C/G  420,413  833  C/C      ...        \n",
       "23     X  148599994   T   C  .   F8  T/C    70,81  151  T/T      ...        \n",
       "24     X  148605752   T   C  .   F8  T/C  336,319  657  T/T      ...        \n",
       "25     X  148685623   G   C  .   F8  G/C    99,80  179  G/G      ...        \n",
       "26     X  148846943   A   T  .   F8  A/T    45,54   99  A/A      ...        \n",
       "27     X  149427266   C   T  .   F8  C/T  278,290  568  C/C      ...        \n",
       "28     X  149429373   A   G  .   F8  A/G  227,241  468  A/A      ...        \n",
       "29     X  149434268   T   C  .   F8  T/C  249,241  490  T/T      ...        \n",
       "\n",
       "    _24  hapA_count hapB_count cumulative_sum_hapa  cumulative_sum_hapb  \\\n",
       "0   374         374        323               374.0                323.0   \n",
       "1   294         294        210               668.0                533.0   \n",
       "2   370         370        303              1038.0                836.0   \n",
       "3   444         444        364              1482.0               1200.0   \n",
       "4   372         372        333              1854.0               1533.0   \n",
       "5   268         335        268              2189.0               1801.0   \n",
       "6   498         498        396              2687.0               2197.0   \n",
       "7   303         440        303              3127.0               2500.0   \n",
       "8   317         317        277              3444.0               2777.0   \n",
       "9   356         478        356              3922.0               3133.0   \n",
       "10  281         404        281              4326.0               3414.0   \n",
       "11  415         415        366              4741.0               3780.0   \n",
       "12  236         303        236              5044.0               4016.0   \n",
       "13  316         406        316              5450.0               4332.0   \n",
       "14  318         395        318              5845.0               4650.0   \n",
       "15  197         263        197              6108.0               4847.0   \n",
       "16  519         519        427              6627.0               5274.0   \n",
       "17  273         273        217              6900.0               5491.0   \n",
       "18  152         152        132              7052.0               5623.0   \n",
       "19  156         253        156              7305.0               5779.0   \n",
       "20  503         503        420              7808.0               6199.0   \n",
       "21  400         495        400              8303.0               6599.0   \n",
       "22  503         556        503              8859.0               7102.0   \n",
       "23  139         201        139              9060.0               7241.0   \n",
       "24  387         527        387              9587.0               7628.0   \n",
       "25  115         173        115                 0.0                  0.0   \n",
       "26  117         149        117               149.0                117.0   \n",
       "27  384         491        384               640.0                501.0   \n",
       "28  316         424        316              1064.0                817.0   \n",
       "29  339         488        339              1552.0               1156.0   \n",
       "\n",
       "   cumulative_ratio cumulative_total upper_boundry lower_boundry        status  \n",
       "0          0.536585            697.0      0.513222      0.486778  Unclassified  \n",
       "1          0.556203           1201.0      0.507673      0.492327  Unclassified  \n",
       "2          0.553895           1874.0      0.504918      0.495082  Unclassified  \n",
       "3          0.552573           2682.0      0.503436      0.496564  Unclassified  \n",
       "4          0.547387           3387.0      0.502721      0.497279  Unclassified  \n",
       "5          0.548622           3990.0      0.502310      0.497690  Unclassified  \n",
       "6          0.550164           4884.0      0.501887      0.498113  Unclassified  \n",
       "7          0.555714           5627.0      0.501638      0.498362  Unclassified  \n",
       "8          0.553609           6221.0      0.501481      0.498519  Unclassified  \n",
       "9          0.555918           7055.0      0.501306      0.498694  Unclassified  \n",
       "10         0.558915           7740.0      0.501191      0.498809  Unclassified  \n",
       "11         0.556390           8521.0      0.501082      0.498918  Unclassified  \n",
       "12         0.556733           9060.0      0.501017      0.498983  Unclassified  \n",
       "13         0.557146           9782.0      0.500942      0.499058  Unclassified  \n",
       "14         0.556932          10495.0      0.500878      0.499122  Unclassified  \n",
       "15         0.557554          10955.0      0.500841      0.499159  Unclassified  \n",
       "16         0.556844          11901.0      0.500774      0.499226  Unclassified  \n",
       "17         0.556856          12391.0      0.500744      0.499256  Unclassified  \n",
       "18         0.556371          12675.0      0.500727      0.499273  Unclassified  \n",
       "19         0.558315          13084.0      0.500704      0.499296  Unclassified  \n",
       "20         0.557436          14007.0      0.500658      0.499342  Unclassified  \n",
       "21         0.557174          14902.0      0.500618      0.499382  Unclassified  \n",
       "22         0.555040          15961.0      0.500577      0.499423  Unclassified  \n",
       "23         0.555794          16301.0      0.500565      0.499435  Unclassified  \n",
       "24         0.556898          17215.0      0.500535      0.499465  Unclassified  \n",
       "25         0.557619          17503.0      0.500527      0.499473          hapA  \n",
       "26         0.560150            266.0      0.534645      0.465355  Unclassified  \n",
       "27         0.560911           1141.0      0.508077      0.491923  Unclassified  \n",
       "28         0.565657           1881.0      0.504899      0.495101  Unclassified  \n",
       "29         0.573117           2708.0      0.503403      0.496597  Unclassified  \n",
       "\n",
       "[30 rows x 33 columns]"
      ]
     },
     "execution_count": 371,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 372,
   "metadata": {},
   "outputs": [],
   "source": [
    "def call_blocks(df):\n",
    "    \n",
    "    block_df = []\n",
    "    \n",
    "    prev_status = 'Unclassified'\n",
    "    snp_count = 0\n",
    "    block_id = 0\n",
    "    first = True\n",
    "    \n",
    "    for snp in df.itertuples():\n",
    "        \n",
    "        if first == True:\n",
    "            \n",
    "            prev_pos = snp.POS\n",
    "            first = False\n",
    "        \n",
    "        status = snp.status\n",
    "        \n",
    "        if status != prev_status:\n",
    "            \n",
    "            block_df.append([block_id, prev_pos, snp.POS, snp_count, status])\n",
    "            \n",
    "            prev_pos = snp.POS\n",
    "            snp_count = 0\n",
    "            block_id = block_id + 1\n",
    "        \n",
    "            \n",
    "        snp_count = snp_count + 1\n",
    "        \n",
    "    return block_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 373,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[0, 148338598, 148685623, 25, 'hapA'],\n",
       " [1, 148685623, 149490984, 25, 'hapA'],\n",
       " [2, 149490984, 149644030, 25, 'hapA'],\n",
       " [3, 149644030, 150197118, 25, 'hapA'],\n",
       " [4, 150197118, 150261734, 25, 'hapA'],\n",
       " [5, 150261734, 150351783, 25, 'hapA'],\n",
       " [6, 150351783, 150439608, 25, 'hapA'],\n",
       " [7, 150439608, 150497850, 25, 'hapA'],\n",
       " [8, 150497850, 150929566, 25, 'hapA'],\n",
       " [9, 150929566, 150998498, 25, 'hapA'],\n",
       " [10, 150998498, 151090219, 25, 'hapA'],\n",
       " [11, 151090219, 151243277, 25, 'hapA'],\n",
       " [12, 151243277, 151370339, 25, 'hapA'],\n",
       " [13, 151370339, 152099691, 25, 'hapA'],\n",
       " [14, 152099691, 152173319, 25, 'hapA'],\n",
       " [15, 152173319, 152587423, 25, 'hapA'],\n",
       " [16, 152587423, 152675931, 25, 'hapA'],\n",
       " [17, 152675931, 152757163, 25, 'hapA'],\n",
       " [18, 152757163, 152932194, 25, 'hapA'],\n",
       " [19, 152932194, 153090060, 25, 'hapA']]"
      ]
     },
     "execution_count": 373,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "call_blocks(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 374,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(516, 33)"
      ]
     },
     "execution_count": 374,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 375,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABVoAAAH1CAYAAAAHwkXPAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAIABJREFUeJzt3WtsnvV9//FPYjs0jdNuWUClEWZpNpdDhozbrmiT6SoTwcYmtrCScBZqGWtF+yApY2UlRCFyA1OniSzQbmxpBT0YsYr1tJZkQ3WXtmwkc6tQAlJbMpRJHNuBHRrj+NoD1PuvrPy5Hb6YOvB6PfJ1X7/b/kn5KvnpnTtX5jRN0wQAAAAAgJds7i96AwAAAAAARzqhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoKhtaJ2amsq6deuyatWqXHzxxdm7d+8h9//+7/8+K1euzLnnnptt27bN2EYBAAAAAGarznYLtm/fnomJiQwPD2d0dDSbNm3KLbfckiR5+umnc9ttt+Xuu+/Os88+mz/8wz/MihUrZnzTAAAAAACzSdtPtO7cuTMDAwNJkr6+vuzevbt1b/78+Xnzm9+cZ599Ns8++2zmzJkzczsFAAAAAJil2n6idWxsLN3d3a3rjo6OTE5OprPz+bcee+yxOfvss3Pw4MFcccUVbX/gO9/5zixZsqSwZQAAAACAV8a+ffty7733tl3XNrR2d3dnfHy8dT01NdWKrCMjI3nsscfyL//yL0mS9773venv788pp5zy//1+S5YsyRe+8IW2GwMAAAAA+EVbuXLltNa1fXRAf39/RkZGkiSjo6Pp7e1t3XvjG9+Y173udZk3b16OOuqoLFy4ME8//fRL3DIAAAAAwJGp7SdaV6xYkR07dmT16tVpmiZDQ0PZunVrenp6Mjg4mG9961s577zzMnfu3PT39+e3f/u3X4l9AwAAAADMGnOapmleyR+4cuVKjw4AAAAAAI4I0+2ZbR8dAAAAAADAixNaAQAAAACKhFYAAAAAgCKhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoEhoBQAAAAAoEloBAAAAAIqEVgAAAACAIqEVAAAAAKBIaAUAAAAAKBJaAQAAAACKhFYAAAAAgCKhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoEhoBQAAAAAoEloBAAAAAIqEVgAAAACAIqEVAAAAAKBIaAUAAAAAKBJaAQAAAACKhFYAAAAAgCKhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoEhoBQAAAAAoEloBAAAAAIqEVgAAAACAIqEVAAAAAKBIaAUAAAAAKBJaAQAAAACKhFYAAAAAgCKhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoEhoBQAAAAAoEloBAAAAAIqEVgAAAACAIqEVAAAAAKBIaAUAAAAAKBJaAQAAAACKhFYAAAAAgCKhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoEhoBQAAAAAoEloBAAAAAIqEVgAAAACAIqEVAAAAAKBIaAUAAAAAKBJaAQAAAACKhFYAAAAAgCKhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoEhoBQAAAAAoEloBAAAAAIo62y2YmprK+vXr8+CDD2bevHnZuHFjjj/++CTJAw88kKGhodba0dHRbNmyJaeffvrM7RgAAAAAYJZpG1q3b9+eiYmJDA8PZ3R0NJs2bcott9ySJDnxxBNz2223JUn++Z//Occcc4zICgAAAAC85rQNrTt37szAwECSpK+vL7t37/65Nfv378/mzZtz++23v/w7BAAAAACY5do+o3VsbCzd3d2t646OjkxOTh6y5s4778xZZ52VRYsWvfw7BAAAAACY5dqG1u7u7oyPj7eup6am0tl56Adhv/SlL+U973nPy787AAAAAIAjQNvQ2t/fn5GRkSTP/2dXvb29h9x/5plnMjExkWOPPXZmdggAAAAAMMu1fUbrihUrsmPHjqxevTpN02RoaChbt25NT09PBgcH86Mf/ShLlix5JfYKAAAAADArtQ2tc+fOzYYNGw55bdmyZa2vTznllNx8880v/84AAAAAAI4QbR8dAAAAAADAixNaAQAAAACKhFYAAAAAgCKhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoEhoBQAAAAAoEloBAAAAAIqEVgAAAACAIqEVAAAAAKBIaAUAAAAAKBJaAQAAAACKhFYAAAAAgCKhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoEhoBQAAAAAoEloBAAAAAIqEVgAAAACAIqEVAAAAAKBIaAUAAAAAKBJaAQAAAACKhFYAAAAAgCKhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoEhoBQAAAAAoEloBAAAAAIqEVgAAAACAIqEVAAAAAKBIaAUAAAAAKBJaAQAAAACKhFYAAAAAgCKhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoEhoBQAAAAAoEloBAAAAAIqEVgAAAACAIqEVAAAAAKBIaAUAAAAAKBJaAQAAAACKhFYAAAAAgCKhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoEhoBQAAAAAoEloBAAAAAIqEVgAAAACAIqEVAAAAAKBIaAUAAAAAKBJaAQAAAACKhFYAAAAAgCKhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoEhoBQAAAAAoEloBAAAAAIo62y2YmprK+vXr8+CDD2bevHnZuHFjjj/++Nb9b3zjG9myZUuS5KSTTsp1112XOXPmzNyOAQAAAABmmbafaN2+fXsmJiYyPDyctWvXZtOmTa17Y2Nj+cu//Mt84hOfyB133JElS5bkxz/+8YxuGAAAAABgtmkbWnfu3JmBgYEkSV9fX3bv3t2695//+Z/p7e3NDTfckAsuuCCLFy/OokWLZm63AAAAAACzUNtHB4yNjaW7u7t13dHRkcnJyXR2dubHP/5x7r333tx11115/etfnwsvvDB9fX1ZunTpYW9k3z2X5rnxfYf9PgAAAACAqq4FS7Lk3Z9+ye9vG1q7u7szPj7eup6amkpn5/Nv+6Vf+qX8xm/8Ro4++ugkydvf/vY88MADLym0Pje+L52vO/qw3wcAAAAAUFX9EGjbRwf09/dnZGQkSTI6Opre3t7WveXLl+ehhx7KU089lcnJyXz3u9/Nr/3ar5U2BAAAAABwpGn7idYVK1Zkx44dWb16dZqmydDQULZu3Zqenp4MDg5m7dq1ed/73pckOeussw4JsQAAAAAArwVtQ+vcuXOzYcOGQ15btmxZ6+uzzz47Z5999su/MwAAAACAI0TbRwcAAAAAAPDihFYAAAAAgCKhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoEhoBQAAAAAoEloBAAAAAIqEVgAAAACAIqEVAAAAAKBIaAUAAAAAKBJaAQAAAACKhFYAAAAAgCKhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoEhoBQAAAAAoEloBAAAAAIqEVgAAAACAIqEVAAAAAKBIaAUAAAAAKBJaAQAAAACKhFYAAAAAgCKhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoEhoBQAAAAAoEloBAAAAAIqEVgAAAACAIqEVAAAAAKBIaAUAAAAAKBJaAQAAAACKhFYAAAAAgCKhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoEhoBQAAAAAoEloBAAAAAIqEVgAAAACAIqEVAAAAAKBIaAUAAAAAKBJaAQAAAACKhFYAAAAAgCKhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoEhoBQAAAAAoEloBAAAAAIqEVgAAAACAIqEVAAAAAKBIaAUAAAAAKBJaAQAAAACKhFYAAAAAgCKhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoEhoBQAAAAAoEloBAAAAAIqEVgAAAACAos52C6amprJ+/fo8+OCDmTdvXjZu3Jjjjz++dX/jxo3ZtWtXFixYkCS5+eabs3DhwpnbMQAAAADALNM2tG7fvj0TExMZHh7O6OhoNm3alFtuuaV1//7778+tt96aRYsWzehGAQAAAABmq7aPDti5c2cGBgaSJH19fdm9e3fr3tTUVPbu3Zt169Zl9erVufPOO2dupwAAAAAAs1TbT7SOjY2lu7u7dd3R0ZHJycl0dnZm//79ueiii3LZZZfl4MGDueSSS7J8+fKccMIJM7ppAAAAAIDZpO0nWru7uzM+Pt66npqaSmfn8312/vz5ueSSSzJ//vx0d3fntNNOy549e2ZutwAAAAAAs1Db0Nrf35+RkZEkyejoaHp7e1v3Hn744VxwwQU5ePBgnnvuuezatSsnn3zyzO0WAAAAAGAWavvogBUrVmTHjh1ZvXp1mqbJ0NBQtm7dmp6engwODuYP/uAPct5556WrqyvnnHNOfv3Xf/2V2DcAAAAAwKzRNrTOnTs3GzZsOOS1ZcuWtb6+/PLLc/nll7/8OwMAAAAAOEK0fXQAAAAAAAAvTmgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoEhoBQAAAAAoEloBAAAAAIqEVgAAAACAIqEVAAAAAKBIaAUAAAAAKBJaAQAAAACKhFYAAAAAgCKhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoEhoBQAAAAAoEloBAAAAAIqEVgAAAACAIqEVAAAAAKBIaAUAAAAAKBJaAQAAAACKhFYAAAAAgCKhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoEhoBQAAAAAoEloBAAAAAIqEVgAAAACAIqEVAAAAAKBIaAUAAAAAKBJaAQAAAACKhFYAAAAAgCKhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoEhoBQAAAAAoEloBAAAAAIqEVgAAAACAIqEVAAAAAKBIaAUAAAAAKBJaAQAAAACKhFYAAAAAgCKhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoEhoBQAAAAAoEloBAAAAAIqEVgAAAACAIqEVAAAAAKBIaAUAAAAAKBJaAQAAAACKhFYAAAAAgCKhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoEhoBQAAAAAoEloBAAAAAIqEVgAAAACAIqEVAAAAAKBIaAUAAAAAKGobWqemprJu3bqsWrUqF198cfbu3fuCa973vvflc5/73IxsEgAAAABgNmsbWrdv356JiYkMDw9n7dq12bRp08+t+eu//uv8z//8z4xsEAAAAABgtmsbWnfu3JmBgYEkSV9fX3bv3n3I/a997WuZM2dOTj/99JnZIQAAAADALNfZbsHY2Fi6u7tb1x0dHZmcnExnZ2ceeuihfPnLX85NN92ULVu2lDbStWBJnhvfV/oeAAAAAAAvRdeCJaX3tw2t3d3dGR8fb11PTU2ls/P5t91111159NFHc+mll2bfvn3p6urKkiVLXtKnW5e8+9OH/R4AAAAAgNmgbWjt7+/PPffck9/7vd/L6Ohoent7W/f+7M/+rPX15s2bs3jxYo8QAAAAAABec9qG1hUrVmTHjh1ZvXp1mqbJ0NBQtm7dmp6engwODr4SewQAAAAAmNXahta5c+dmw4YNh7y2bNmyn1v3wQ9+8OXbFQAAAADAEWTuL3oDAAAAAABHOqEVAAAAAKBIaAUAAAAAKBJaAQAAAACKhFYAAAAAgCKhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoEhoBQAAAAAoEloBAAAAAIqEVgAAAACAIqEVAAAAAKBIaAUAAAAAKBJaAQAAAACKhFYAAAAAgCKhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoEhoBQAAAAAoEloBAAAAAIqEVgAAAACAIqEVAAAAAKBIaAUAAAAAKBJaAQAAAACKhFYAAAAAgCKhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoEhoBQAAAAAoEloBAAAAAIqEVgAAAACAIqEVAAAAAKBIaAUAAAAAKBJaAQAAAACKhFYAAAAAgCKhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoEhoBQAAAAAoEloBAAAAAIqEVgAAAACAIqEVAAAAAKBIaAUAAAAAKBJaAQAAAACKhFYAAAAAgCKhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoEhoBQAAAAAoEloBAAAAAIqEVgAAAACAIqEVAAAAAKBIaAUAAAAAKBJaAQAAAACKhFYAAAAAgCKhFQAAAACgSGgFAAAAACgSWgEAAAAAioRWAAAAAIAioRUAAAAAoKhtaJ2amsq6deuyatWqXHzxxdm7d+8h9z/zmc/k3HPPzR//8R/nnnvumbGNAgAAAADMVp3tFmzfvj0TExMZHh7O6OhoNm3alFtuuSVJ8tRTT+Wzn/1s7rrrrhw4cCBnn312fud3fidz5syZ8Y0DAAAAAMwWbT/RunPnzgwMDCRJ+vr6snv37ta9RYsW5Z/+6Z/S1dWVJ554Im94wxtEVgAAAADgNadtaB0bG0t3d3fruqOjI5OTk63rzs7O3H777Vm1alXOPPPMmdklAAAAAMAs1ja0dnd3Z3x8vHU9NTWVzs5Dnzhw0UUX5Zvf/Gb+4z/+I9/5znde/l0CAAAAAMxibUNrf39/RkZGkiSjo6Pp7e1t3fvhD3+YK6+8Mk3TpKurK/PmzcvcuW2/JQAAAADAq0rb/wxrxYoV2bFjR1avXp2maTI0NJStW7emp6cng4ODOeGEE7Jq1arMmTMnAwMD+c3f/M1XYt8AAAAAALPGnKZpmlfyB65cuTJf+MIXXskfCQAAAADwkky3Z/p3/gAAAAAARUIrAAAAAECR0AoAAAAAUCS0AgAAAAAUCa0AAAAAAEVCKwAAAABAkdAKAAAAAFAktAIAAAAAFAmtAAAAAABFQisAAAAAQJHQCgAAAABQJLQCAAAAABQJrQAAAAAARUIrAAAAAECR0AoAAAAAUCS0AgAAAAAUCa0AAAAAAEVCKwAAAABAkdAKAAAAAFAktAIAAAAAFAmtAAAAAABFQisAAAAAQJHQCgAAAABQJLQCAAAAABQJrQAAAAAARUIrAAAAAECR0AoAAAAAUCS0AgAAAAAUCa0AAAAAAEVCKwAAAABAkdAKAAAAAFAktAIAAAAAFAmtAAAAAABFQisAAAAAQJHQCgAAAABQJLQCAAAAABQJrQAAAAAARUIrAAAAAECR0AoAAAAAUCS0AgAAAAAUCa0AAAAAAEVCKwAAAABAkdAKAAAAAFAktAIAAAAAFAmtAAAAAABFQisAAAAAQJHQCgAAAABQJLQCAAAAABQJrQAAAAAARUIrAAAAAECR0AoAAAAAUCS0AgAAAAAUCa0AAAAAAEVCKwAAAABAkdAKAAAAAFAktAIAAAAAFAmtAAAAAABFQisAAAAAQJHQCgAAAABQJLQCAAAAABQJrQAAAAAARUIrAAAAAECR0AoAAAAAUCS0AgAAAAAUCa0AAAAAAEVCKwAAAABAkdAKAAAAAFAktAIAAAAAFAmtAAAAAABFQisAAAAAQFFnuwVTU1NZv359HnzwwcybNy8bN27M8ccf37r/qU99Kl/5yleSJO9617ty5ZVXztxuAQAAAABmobafaN2+fXsmJiYyPDyctWvXZtOmTa17jzzySL74xS/m85//fIaHh/Nv//Zv2bNnz4xuGAAAAABgtmn7idadO3dmYGAgSdLX15fdu3e37r3pTW/Krbfemo6OjiTJ5ORkjjrqqBnaKgAAAADA7NQ2tI6NjaW7u7t13dHRkcnJyXR2dqarqyuLFi1K0zS58cYbc9JJJ2Xp0qUv+v327duXlStX1ncOAAAAADDD9u3bN611bUNrd3d3xsfHW9dTU1Pp7Px/bztw4ECuueaaLFiwINddd13bH3jvvfdOa2MAAAAAAEeKts9o7e/vz8jISJJkdHQ0vb29rXtN0+QDH/hA3vrWt2bDhg2tRwgAAAAAALyWzGmapnmxBVNTU1m/fn0eeuihNE2ToaGhjIyMpKenJ1NTU1mzZk36+vpa69esWZNTTz11xjcOAAAAADBbtA2tAAAAAAC8uLaPDgAAAAAA4MUJrQAAAAAARUIrAAAAAEBR5y96A69WBw8ezEc/+tH86Ec/SkdHRz72sY+lp6cnSTI0NJSlS5fm/PPPT5J86lOfyle+8pUkybve9a5ceeWVaZomp59+en71V381SdLX15e1a9fmX//1X7Nly5Z0dnbm3HPPzXnnnZef/vSnueqqq/Lkk09mwYIFueGGG7Jo0aLDWsvsdzgztXHjxuzatSsLFixIktx888157rnn8uEPfzg//elPc8wxx+RjH/tY5s+fnzvuuCOf//zn09nZmfe///1597vfnaeeeqq8ltltuvP0wAMPZGhoqPW+0dHRbNmyJaecckrOPPPM9Pb2JknOOOOMXHrppebpNeyFZmp8fDzXX399Ojo6Mm/evNxwww1ZvHjxjM2JmXr1OJx5co5iOg5nppyjaGe68/T44487RzEtLzRTExMTufbaa9M0TU444YRce+216ejocI6ircOZJ+eoGdAwI7Zt29b8+Z//edM0TfOd73yn+dM//dPmySefbN773vc2g4ODzWc/+9mmaZrmv/7rv5o/+qM/aiYnJ5uDBw82q1atah544IHm4Ycfbq644opDvufExERzxhlnND/5yU+aAwcONCtXrmwee+yx5h/+4R+am266qWmapvnyl7/cXH/99Ye1liPDdGeqaZpm9erVzZNPPnnI+6+//vrmH//xH5umaZpPfvKTzdatW5vHHnus+f3f//3mwIEDzdNPP936urqW2e9w5ulnvvrVrzZr1qxpmqZpduzY0WzYsOGQ++bpte2FZurCCy9svv/97zdN0zSf+9znmqGhoRmbEzP16jLdeXKOYrqmO1NN4xxFe4czTz/jHMWLeaGZev/739/8+7//e9M0TXP11Vc3d999t3MU0zLdeXKOmhkeHTBDzjjjjFx//fVJkv/+7//O4sWLMz4+ng9+8IM555xzWuve9KY35dZbb01HR0fmzp2bycnJHHXUUbn//vvz6KOP5uKLL87ll1+eH/7wh/nBD36Qnp6evPGNb8y8efPytre9Lffdd1927tyZgYGBJMnpp5+eb3/724e1liPDdGdqamoqe/fuzbp167J69erceeedSfJzv/bf+ta38r3vfS+nnnpq5s2bl4ULF6anpyd79uwpr2X2m+48/cz+/fuzefPm/MVf/EWSZPfu3bn//vtz0UUX5UMf+lAee+wx8/Qa90Iz9Vd/9Vc58cQTkzz/N+tHHXXUjM2JmXp1me48OUcxXdOdKecopmO68/QzzlG080IztXnz5rzjHe/IxMREHn/88fzKr/yKcxTTMt15co6aGR4dMIM6Oztz9dVXZ9u2bbnpppty3HHH5bjjjsvIyEhrTVdXVxYtWpSmaXLjjTfmpJNOytKlS/PEE0/kT/7kT/K7v/u7ue+++3LVVVflIx/5SBYuXNh674IFCzI2NpaxsbHW6wsWLMgzzzxzyGvt1nLkmM5M7d+/PxdddFEuu+yyHDx4MJdcckmWL19enhMz9eoznXn6mTvvvDNnnXVW6592vOUtb8ny5cvzW7/1W/niF7+YjRs3ZnBw0Dy9xv3fmTrmmGOSJLt27crtt9+ez3zmM/nmN785I3Nipl59pjNPzlEcjunMlHMU0zWdefoZ5yim4//OVEdHR/bt25fLLrss3d3dWbp0aR555BHnKKZlOvPkHDUzfKJ1ht1www35+te/nmuvvTb79+9/wTUHDhzIhz/84YyPj+e6665LkixfvjyDg4NJkre//e159NFH093dnfHx8db7xsfHs3DhwkNeHx8fzxve8IbDWsuRpd1MzZ8/P5dccknmz5+f7u7unHbaadmzZ095TszUq9N0fo9Kki996Ut5z3ve07o+7bTT8s53vjNJsmLFinz/+983TyT5+Zn66le/muuuuy5/+7d/m0WLFs3YnJipV6d285Q4R3F42s2UcxSHYzq/RyXOUUzf/52pJUuW5O67787555+fTZs2OUdxWNrNU+IcNROE1hly11135ZOf/GSS58PXnDlz0tHR8XPrmqbJBz7wgbz1rW/Nhg0bWmv+5m/+Jp/+9KeTJHv27Mmb3/zmLFu2LHv37s1PfvKTTExM5L777supp56a/v7+fOMb30iSjIyM5G1ve9threXIMN2Zevjhh3PBBRfk4MGDee6557Jr166cfPLJL/hrf8opp2Tnzp05cOBAnnnmmfzgBz9Ib29veS2z33TnKUmeeeaZTExM5Nhjj2299tGPfjRf//rXkyTf/va3c/LJJ5un17gXmqlt27bl9ttvz2233ZbjjjsuSWZsTszUq8t058k5iuma7kw5RzEd052nxDmK6Xmhmbryyivz8MMPJ3n+039z5851jmJapjtPzlEzY07TNM0vehOvRvv3789HPvKRPPHEE5mcnMzll1+eM844I0myefPmLF68OOeff362bduWNWvWpK+vr/XeNWvW5C1veUuuuuqq7N+/Px0dHVm3bl2WLVvW+p/bmqbJueeemwsvvDDPPvtsrr766jz++OPp6urKxz/+8Rx99NHfM6v2AAABEklEQVSHtZbZb7ozlSR/93d/l6997Wvp6urKOeeck/PPPz9PPPFErr766oyPj+eXf/mX8/GPfzyvf/3rc8cdd2R4eDhN0+SKK67ImWee+bKsZXY7nHn63ve+l0984hO5+eabW+9/5JFHcs011yR5/g/vjRs35phjjjFPr2EvNFPXXHNNjj322NbfVr/jHe/Ihz70oRmbEzP16jHdeTrxxBOdo5iWw/k9yjmKdg5nnpyjmI4XmqlFixblxhtvTFdX1ysyJ2bq1WO68/Td737XOWoGCK0AAAAAAEUeHQAAAAAAUCS0AgAAAAAUCa0AAAAAAEVCKwAAAABAkdAKAAAAAFAktAIAAAAAFAmtAAAAAABF/wvImp+ygeOYwQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1332x468 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig = plt.figure()\n",
    "ax = fig.add_axes([0,0,1,1])\n",
    "fig.set_size_inches(18.5, 6.5)\n",
    "\n",
    "ax.set_xlim([31115694,33357658])\n",
    "ax.get_xaxis().get_major_formatter().set_scientific(False)\n",
    "ax.set_ylim([0.2,0.8])\n",
    "rect = patches.Rectangle((31115694+10000,0.4),(33357658-31115694)-20000,0.20, color='goldenrod', alpha=0.75)\n",
    "\n",
    "ax.add_patch(rect)\n",
    "\n",
    "for x in call_blocks(df):\n",
    "    \n",
    "    if x[4] == 'hapA':\n",
    "\n",
    "        ax.arrow(x[1], 0.65, (x[2]-x[1]), 0, head_width=0.02, head_length=7500, color='red')\n",
    "        \n",
    "    else:\n",
    "        \n",
    "        ax.arrow(x[1], 0.65, (x[2]-x[1]), 0, head_width=0.02, head_length=7500, color='blue')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 376,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[0, 153828373, 153090060, 15, 'hapA'],\n",
       " [1, 153090060, 152932194, 25, 'hapA'],\n",
       " [2, 152932194, 152757163, 25, 'hapA'],\n",
       " [3, 152757163, 152675931, 25, 'hapA'],\n",
       " [4, 152675931, 152587423, 25, 'hapA'],\n",
       " [5, 152587423, 152173319, 25, 'hapA'],\n",
       " [6, 152173319, 152099691, 25, 'hapA'],\n",
       " [7, 152099691, 151370339, 25, 'hapA'],\n",
       " [8, 151370339, 151243277, 25, 'hapA'],\n",
       " [9, 151243277, 151090219, 25, 'hapA'],\n",
       " [10, 151090219, 150998498, 25, 'hapA'],\n",
       " [11, 150998498, 150929566, 25, 'hapA'],\n",
       " [12, 150929566, 150497850, 25, 'hapA'],\n",
       " [13, 150497850, 150439608, 25, 'hapA'],\n",
       " [14, 150439608, 150351783, 25, 'hapA'],\n",
       " [15, 150351783, 150261734, 25, 'hapA'],\n",
       " [16, 150261734, 150197118, 25, 'hapA'],\n",
       " [17, 150197118, 149644030, 25, 'hapA'],\n",
       " [18, 149644030, 149490984, 25, 'hapA'],\n",
       " [19, 149490984, 148685623, 25, 'hapA']]"
      ]
     },
     "execution_count": 376,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "call_blocks(df.sort_values('POS', ascending=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
